{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8d235812",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting Bio\n",
      "  Downloading bio-1.5.9-py3-none-any.whl (276 kB)\n",
      "     ---------------------------------------- 0.0/276.4 kB ? eta -:--:--\n",
      "     -------- ------------------------------ 61.4/276.4 kB 1.1 MB/s eta 0:00:01\n",
      "     -------------------------------- ----- 235.5/276.4 kB 2.4 MB/s eta 0:00:01\n",
      "     -------------------------------------- 276.4/276.4 kB 2.1 MB/s eta 0:00:00\n",
      "Collecting pooch\n",
      "  Downloading pooch-1.7.0-py3-none-any.whl (60 kB)\n",
      "     ---------------------------------------- 0.0/60.9 kB ? eta -:--:--\n",
      "     ---------------------------------------- 60.9/60.9 kB ? eta 0:00:00\n",
      "Requirement already satisfied: tqdm in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from Bio) (4.64.1)\n",
      "Collecting gprofiler-official\n",
      "  Downloading gprofiler_official-1.0.0-py3-none-any.whl (9.3 kB)\n",
      "Requirement already satisfied: requests in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from Bio) (2.28.2)\n",
      "Requirement already satisfied: pandas in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from Bio) (1.5.3)\n",
      "Collecting mygene\n",
      "  Downloading mygene-3.2.2-py2.py3-none-any.whl (5.4 kB)\n",
      "Collecting biopython>=1.80\n",
      "  Downloading biopython-1.81-cp310-cp310-win_amd64.whl (2.7 MB)\n",
      "     ---------------------------------------- 0.0/2.7 MB ? eta -:--:--\n",
      "     ----- ---------------------------------- 0.4/2.7 MB 7.8 MB/s eta 0:00:01\n",
      "     ----------------- ---------------------- 1.2/2.7 MB 12.5 MB/s eta 0:00:01\n",
      "     ------------------------------ --------- 2.1/2.7 MB 14.7 MB/s eta 0:00:01\n",
      "     ---------------------------------------- 2.7/2.7 MB 14.4 MB/s eta 0:00:00\n",
      "Requirement already satisfied: numpy in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from biopython>=1.80->Bio) (1.23.1)\n",
      "Collecting biothings-client>=0.2.6\n",
      "  Downloading biothings_client-0.3.0-py2.py3-none-any.whl (29 kB)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from pandas->Bio) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from pandas->Bio) (2022.7.1)\n",
      "Requirement already satisfied: platformdirs>=2.5.0 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from pooch->Bio) (3.0.0)\n",
      "Requirement already satisfied: packaging>=20.0 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from pooch->Bio) (23.0)\n",
      "Requirement already satisfied: idna<4,>=2.5 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->Bio) (3.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->Bio) (2022.12.7)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->Bio) (1.26.14)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from requests->Bio) (3.0.1)\n",
      "Requirement already satisfied: colorama in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from tqdm->Bio) (0.4.4)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\paulius\\appdata\\local\\packages\\pythonsoftwarefoundation.python.3.10_qbz5n2kfra8p0\\localcache\\local-packages\\python310\\site-packages (from python-dateutil>=2.8.1->pandas->Bio) (1.16.0)\n",
      "Installing collected packages: biopython, pooch, gprofiler-official, biothings-client, mygene, Bio\n",
      "Successfully installed Bio-1.5.9 biopython-1.81 biothings-client-0.3.0 gprofiler-official-1.0.0 mygene-3.2.2 pooch-1.7.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  WARNING: The scripts bio.exe and fasta_filter.exe are installed in 'C:\\Users\\Paulius\\AppData\\Local\\Packages\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\LocalCache\\local-packages\\Python310\\Scripts' which is not on PATH.\n",
      "  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.\n",
      "\n",
      "[notice] A new release of pip is available: 23.0.1 -> 23.3\n",
      "[notice] To update, run: C:\\Users\\Paulius\\AppData\\Local\\Microsoft\\WindowsApps\\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\\python.exe -m pip install --upgrade pip\n"
     ]
    }
   ],
   "source": [
    "!pip install Bio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "e19ce938",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "import math\n",
    "from scipy.spatial.distance import pdist, squareform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "45167914",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "def extract_valid_start_stop_pairs(sequence):\n",
    "    start_codon = \"ATG\"\n",
    "    stop_codons = [\"TAA\", \"TAG\", \"TGA\"]\n",
    "    \n",
    "    valid_pairs = []\n",
    "    \n",
    "    def get_pairs(seq):\n",
    "        pairs = []\n",
    "        for i in range(len(seq) - 2):\n",
    "            if seq[i:i+3] == start_codon:\n",
    "                for j in range(i+3, len(seq) - 2, 3):\n",
    "                    if seq[j:j+3] in stop_codons and not any(stop in seq[i+3:j] for stop in stop_codons):\n",
    "                        pairs.append((i, j))\n",
    "                        break\n",
    "        return pairs\n",
    "    \n",
    "    valid_pairs.extend(get_pairs(sequence))\n",
    "    \n",
    "    reverse_seq = str(Seq(sequence).reverse_complement())\n",
    "    valid_pairs.extend(get_pairs(reverse_seq))\n",
    "\n",
    "    return valid_pairs\n",
    "#returns whole fragment\n",
    "def filter_long_fragments(sequence, pairs, max_length=100):\n",
    "    fragments = []\n",
    "    for start, stop in pairs:\n",
    "        fragment = sequence[start:stop+3]\n",
    "        if len(fragment) <= max_length:\n",
    "            fragments.append(fragment)\n",
    "    return fragments\n",
    "#returns start and finish pairs\n",
    "def filter_long_fragment_pairs(sequence, pairs, max_length=100):\n",
    "    positions = []\n",
    "    for start, stop in pairs:\n",
    "        fragment_length = stop + 3 - start\n",
    "        if fragment_length <= max_length:\n",
    "            positions.append((start, stop))\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "168ad889",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_protein(dna_sequence):\n",
    "    dna_seq = Seq(dna_sequence)\n",
    "    protein_seq = dna_seq.translate()\n",
    "    return protein_seq\n",
    "def extract_dikodons(protein_sequence):\n",
    "    dikodons = [protein_sequence[i:i+2] for i in range(0, len(protein_sequence)-1)]\n",
    "    return dikodons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "aad5b146",
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_frequencies(sequence):\n",
    "    codons = all_possible_codons()\n",
    "    frequencies = {codon: 0 for codon in codons}\n",
    "    \n",
    "    for i in range(0, len(sequence)-2, 3):\n",
    "        codon = sequence[i:i+3]\n",
    "        if codon in frequencies:\n",
    "            frequencies[codon] += 1\n",
    "\n",
    "    total_codons = sum(frequencies.values())\n",
    "    for codon, count in frequencies.items():\n",
    "        frequencies[codon] = count / total_codons\n",
    "\n",
    "    return frequencies\n",
    "def dikodon_frequencies(sequence):\n",
    "    dikodons = all_possible_dikodons()\n",
    "    frequencies = {dikodon: 0 for dikodon in dikodons}\n",
    "    \n",
    "    for i in range(0, len(sequence)-5, 3):\n",
    "        dikodon = sequence[i:i+6]\n",
    "        if dikodon in frequencies:\n",
    "            frequencies[dikodon] += 1\n",
    "\n",
    "    total_dikodons = sum(frequencies.values())\n",
    "    for dikodon, count in frequencies.items():\n",
    "        frequencies[dikodon] = count / total_dikodons \n",
    "\n",
    "    return frequencies\n",
    "\n",
    "def all_possible_codons():\n",
    "    bases = 'ATGC'\n",
    "    return [a+b+c for a in bases for b in bases for c in bases]\n",
    "\n",
    "def all_possible_dikodons():\n",
    "    codons = all_possible_codons()\n",
    "    return [a+b for a in codons for b in codons]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "7c915c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "def amino_acid_frequencies(protein_sequences):\n",
    "    amino_acids = \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "    amino_acid_counts = {aa: 0 for aa in amino_acids}\n",
    "\n",
    "    for protein in protein_sequences:\n",
    "        for aa in protein:\n",
    "            if aa in amino_acid_counts:\n",
    "                amino_acid_counts[aa] += 1\n",
    "\n",
    "    total_amino_acids = sum(amino_acid_counts.values())\n",
    "\n",
    "    amino_acid_freqs = {aa: count / total_amino_acids for aa, count in amino_acid_counts.items()}\n",
    "    return amino_acid_freqs\n",
    "def print_amino_acid_frequencies(protein_sequences):\n",
    "    amino_acid_freqs = amino_acid_frequencies(protein_sequences)\n",
    "    header = \"|Amino Acid | Frequency|\"\n",
    "    print(\"-\" * len(header))\n",
    "    print(header)\n",
    "    print(\"-\" * len(header))\n",
    "    for aa, freq in amino_acid_freqs.items():\n",
    "        print(f\"|{aa:^11}| {freq:.6f} |\")\n",
    "    print(\"-\" * len(header))\n",
    "\n",
    "def dikodon_frequency_matrix(protein_sequences):\n",
    "    amino_acids = \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "    \n",
    "    matrix = {(aa1, aa2): 0 for aa1 in amino_acids for aa2 in amino_acids}\n",
    "    total_dikodons = 0\n",
    "\n",
    "    for protein in protein_sequences:\n",
    "        for i in range(len(protein) - 1):\n",
    "            dikodon = (protein[i], protein[i+1])\n",
    "            if dikodon in matrix:\n",
    "                matrix[dikodon] += 1\n",
    "                total_dikodons += 1\n",
    "\n",
    "    for dikodon, count in matrix.items():\n",
    "        matrix[dikodon] = count / total_dikodons\n",
    "    \n",
    "    return matrix\n",
    "    \n",
    "def print_dikodon_frequency_matrix(protein_sequences):\n",
    "    amino_acids = \"ACDEFGHIKLMNPQRSTVWY\"\n",
    "    matrix = dikodon_frequency_matrix(protein_sequences)\n",
    "    print(\"-\"*190)\n",
    "    print(\"|\" + \" \" * 8 + \"|\", \" | \".join([f\"{aa:^6}\" for aa in amino_acids]), \"|\")\n",
    "    print(\"-\" * 190)\n",
    "\n",
    "    for aa1 in amino_acids:\n",
    "        row_data = [\"| \" + f\"{matrix[(aa1, aa2)]:>6.4f} \" for aa2 in amino_acids]\n",
    "        print(f\"| {aa1:^6}\", \"\".join(row_data), \"|\")\n",
    "\n",
    "    print(\"-\" * 190)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "d00e1fd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_results(file_name):\n",
    "    sequence_record = SeqIO.read(file_name, \"fasta\")\n",
    "    sequence = sequence_record.seq\n",
    "\n",
    "    pairs = extract_valid_start_stop_pairs(sequence)\n",
    "    #print(\"\\n\")\n",
    "\n",
    "\n",
    "    filtered_fragments = filter_long_fragment_pairs(sequence, pairs)\n",
    "    #print(\"\\n\")\n",
    "\n",
    "    protein_sequences = []\n",
    "    for start, stop in filtered_fragments:\n",
    "        dna_fragment = sequence[start:stop+3]  \n",
    "        protein_sequence = convert_to_protein(dna_fragment)\n",
    "        protein_sequences.append(protein_sequence)\n",
    "        dikodons = extract_dikodons(protein_sequence)\n",
    "        #print(f\"\\nDNA Fragment: {dna_fragment}\")\n",
    "        #print(f\"Protein sequence: {protein_sequence}\")\n",
    "        #print(f\"Dikodons: {dikodons}\")\n",
    "    print_amino_acid_frequencies(protein_sequences)\n",
    "    print_dikodon_frequency_matrix(protein_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "5c729935",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Codon Distance Matrix\n",
      "8\n",
      "coronavirus 0.000 0.111 0.067 0.159 0.035 0.068 0.065 0.087\n",
      "adenovirus 0.111 0.000 0.174 0.053 0.141 0.043 0.116 0.184\n",
      "U18337.1 0.067 0.174 0.000 0.223 0.042 0.132 0.111 0.082\n",
      "herpesvirus 0.159 0.053 0.223 0.000 0.191 0.093 0.164 0.231\n",
      "Lactococcus_phage 0.035 0.141 0.042 0.191 0.000 0.098 0.069 0.067\n",
      "KM389305.1 0.068 0.043 0.132 0.093 0.098 0.000 0.082 0.143\n",
      "NC_028697.1 0.065 0.116 0.111 0.164 0.069 0.082 0.000 0.081\n",
      "KC821626.1 0.087 0.184 0.082 0.231 0.067 0.143 0.081 0.000\n",
      "Dikodon Distance Matrix\n",
      "8\n",
      "coronavirus 0.000 0.019 0.023 0.023 0.020 0.020 0.020 0.025\n",
      "adenovirus 0.019 0.000 0.023 0.017 0.020 0.016 0.018 0.026\n",
      "U18337.1 0.023 0.023 0.000 0.025 0.019 0.022 0.022 0.024\n",
      "herpesvirus 0.023 0.017 0.025 0.000 0.024 0.018 0.022 0.030\n",
      "Lactococcus_phage 0.020 0.020 0.019 0.024 0.000 0.020 0.018 0.023\n",
      "KM389305.1 0.020 0.016 0.022 0.018 0.020 0.000 0.019 0.027\n",
      "NC_028697.1 0.020 0.018 0.022 0.022 0.018 0.019 0.000 0.023\n",
      "KC821626.1 0.025 0.026 0.024 0.030 0.023 0.027 0.023 0.000\n"
     ]
    }
   ],
   "source": [
    "def calculate_amino_acid_dikodon_frequencies(dna_sequence):\n",
    "    protein_sequence = convert_to_protein(dna_sequence)\n",
    "\n",
    "    amino_acid_freqs = amino_acid_frequencies(protein_sequence)  # Assuming you have a function for this\n",
    "    dikodon_freqs = dikodon_frequencies(protein_sequence)\n",
    "    \n",
    "    dikodon_format_conversion = {(dikodon[0], dikodon[1]): freq for dikodon, freq in dikodon_freqs.items()}\n",
    "    \n",
    "    return amino_acid_freqs, dikodon_format_conversion\n",
    "\n",
    "def euclidean_distance(freqs1, freqs2):\n",
    "    dist = 0\n",
    "    for k in freqs1.keys():\n",
    "        dist += (freqs1[k] - freqs2[k]) ** 2\n",
    "    return math.sqrt(dist)\n",
    "\n",
    "def compute_distance_matrices(fasta_files):\n",
    "    sequences = {}\n",
    "    for file in fasta_files:\n",
    "        with open(file, 'r') as f:\n",
    "            for record in SeqIO.parse(f, 'fasta'):\n",
    "                sequences[record.id] = str(record.seq)\n",
    "    \n",
    "    freqs = {id: calculate_codon_dikodon_frequencies(seq) for id, seq in sequences.items()}\n",
    "    \n",
    "    distance_matrix_codon = {}\n",
    "    distance_matrix_dikodon = {}\n",
    "    for id1 in sequences:\n",
    "        distance_matrix_codon[id1] = {}\n",
    "        distance_matrix_dikodon[id1] = {}\n",
    "        for id2 in sequences:\n",
    "            dist_codon = euclidean_distance(freqs[id1][0], freqs[id2][0])\n",
    "            dist_dikodon = euclidean_distance(freqs[id1][1], freqs[id2][1])\n",
    "            \n",
    "            distance_matrix_codon[id1][id2] = dist_codon\n",
    "            distance_matrix_dikodon[id1][id2] = dist_dikodon\n",
    "    \n",
    "    return distance_matrix_codon, distance_matrix_dikodon\n",
    "\n",
    "def print_phylip_format(distance_matrix, title):\n",
    "    ids = list(distance_matrix.keys())\n",
    "    n = len(ids)\n",
    "    \n",
    "    print(title)\n",
    "    print(n)\n",
    "    for id1 in ids:\n",
    "        row = [id1] + [\"{:.3f}\".format(distance_matrix[id1][id2]) for id2 in ids]\n",
    "        print(\" \".join(row))\n",
    "\n",
    "fasta_files = [\"data\\\\mamalian1.fasta\", \"data\\\\mamalian2.fasta\", \"data\\\\mamalian3.fasta\", \n",
    "               \"data\\\\mamalian4.fasta\", \"data\\\\bacterial1.fasta\", \"data\\\\bacterial2.fasta\",\n",
    "               \"data\\\\bacterial3.fasta\", \"data\\\\bacterial4.fasta\"]\n",
    "distance_matrix_codon, distance_matrix_dikodon = compute_distance_matrices(fasta_files)\n",
    "print_phylip_format(distance_matrix_codon, \"Codon Distance Matrix\")\n",
    "print_phylip_format(distance_matrix_dikodon, \"Dikodon Distance Matrix\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "aab0bb50",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.031453 |\n",
      "|     C     | 0.042436 |\n",
      "|     D     | 0.016975 |\n",
      "|     E     | 0.022966 |\n",
      "|     F     | 0.050924 |\n",
      "|     G     | 0.031952 |\n",
      "|     H     | 0.047928 |\n",
      "|     I     | 0.045931 |\n",
      "|     K     | 0.043435 |\n",
      "|     L     | 0.127309 |\n",
      "|     M     | 0.127808 |\n",
      "|     N     | 0.039940 |\n",
      "|     P     | 0.025462 |\n",
      "|     Q     | 0.045931 |\n",
      "|     R     | 0.039940 |\n",
      "|     S     | 0.069396 |\n",
      "|     T     | 0.060409 |\n",
      "|     V     | 0.072391 |\n",
      "|     W     | 0.017474 |\n",
      "|     Y     | 0.039940 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0013 | 0.0006 | 0.0006 | 0.0000 | 0.0013 | 0.0025 | 0.0019 | 0.0031 | 0.0013 | 0.0044 | 0.0019 | 0.0025 | 0.0000 | 0.0038 | 0.0006 | 0.0013 | 0.0019 | 0.0025 | 0.0000 | 0.0006  |\n",
      "|   C    | 0.0006 | 0.0019 | 0.0006 | 0.0019 | 0.0019 | 0.0013 | 0.0013 | 0.0006 | 0.0006 | 0.0082 | 0.0019 | 0.0031 | 0.0013 | 0.0006 | 0.0038 | 0.0013 | 0.0057 | 0.0025 | 0.0006 | 0.0038  |\n",
      "|   D    | 0.0013 | 0.0006 | 0.0000 | 0.0000 | 0.0000 | 0.0013 | 0.0013 | 0.0019 | 0.0006 | 0.0025 | 0.0000 | 0.0000 | 0.0006 | 0.0000 | 0.0031 | 0.0019 | 0.0013 | 0.0006 | 0.0006 | 0.0013  |\n",
      "|   E    | 0.0006 | 0.0000 | 0.0006 | 0.0013 | 0.0019 | 0.0006 | 0.0006 | 0.0013 | 0.0013 | 0.0013 | 0.0025 | 0.0013 | 0.0031 | 0.0006 | 0.0006 | 0.0019 | 0.0000 | 0.0038 | 0.0000 | 0.0013  |\n",
      "|   F    | 0.0019 | 0.0044 | 0.0013 | 0.0000 | 0.0044 | 0.0006 | 0.0025 | 0.0006 | 0.0019 | 0.0057 | 0.0013 | 0.0000 | 0.0038 | 0.0019 | 0.0025 | 0.0038 | 0.0019 | 0.0063 | 0.0019 | 0.0057  |\n",
      "|   G    | 0.0006 | 0.0013 | 0.0006 | 0.0000 | 0.0013 | 0.0013 | 0.0013 | 0.0013 | 0.0025 | 0.0025 | 0.0000 | 0.0025 | 0.0013 | 0.0006 | 0.0025 | 0.0019 | 0.0038 | 0.0031 | 0.0000 | 0.0044  |\n",
      "|   H    | 0.0025 | 0.0050 | 0.0000 | 0.0000 | 0.0038 | 0.0000 | 0.0044 | 0.0019 | 0.0019 | 0.0101 | 0.0013 | 0.0025 | 0.0006 | 0.0031 | 0.0019 | 0.0063 | 0.0025 | 0.0013 | 0.0000 | 0.0006  |\n",
      "|   I    | 0.0000 | 0.0019 | 0.0000 | 0.0006 | 0.0006 | 0.0013 | 0.0057 | 0.0019 | 0.0031 | 0.0075 | 0.0006 | 0.0019 | 0.0006 | 0.0038 | 0.0000 | 0.0038 | 0.0050 | 0.0031 | 0.0006 | 0.0019  |\n",
      "|   K    | 0.0006 | 0.0019 | 0.0006 | 0.0013 | 0.0006 | 0.0000 | 0.0038 | 0.0031 | 0.0050 | 0.0075 | 0.0019 | 0.0013 | 0.0025 | 0.0000 | 0.0038 | 0.0031 | 0.0025 | 0.0050 | 0.0006 | 0.0019  |\n",
      "|   L    | 0.0063 | 0.0050 | 0.0013 | 0.0019 | 0.0094 | 0.0038 | 0.0069 | 0.0006 | 0.0044 | 0.0276 | 0.0019 | 0.0044 | 0.0044 | 0.0094 | 0.0031 | 0.0088 | 0.0019 | 0.0088 | 0.0031 | 0.0025  |\n",
      "|   M    | 0.0044 | 0.0063 | 0.0013 | 0.0063 | 0.0038 | 0.0063 | 0.0082 | 0.0082 | 0.0057 | 0.0214 | 0.0019 | 0.0038 | 0.0038 | 0.0069 | 0.0063 | 0.0101 | 0.0069 | 0.0163 | 0.0019 | 0.0044  |\n",
      "|   N    | 0.0031 | 0.0013 | 0.0019 | 0.0006 | 0.0019 | 0.0013 | 0.0019 | 0.0025 | 0.0006 | 0.0044 | 0.0006 | 0.0031 | 0.0006 | 0.0000 | 0.0025 | 0.0038 | 0.0019 | 0.0019 | 0.0000 | 0.0025  |\n",
      "|   P    | 0.0000 | 0.0013 | 0.0000 | 0.0013 | 0.0031 | 0.0006 | 0.0000 | 0.0031 | 0.0013 | 0.0038 | 0.0006 | 0.0006 | 0.0000 | 0.0006 | 0.0025 | 0.0006 | 0.0006 | 0.0019 | 0.0000 | 0.0006  |\n",
      "|   Q    | 0.0025 | 0.0031 | 0.0000 | 0.0050 | 0.0006 | 0.0013 | 0.0031 | 0.0044 | 0.0044 | 0.0069 | 0.0025 | 0.0019 | 0.0006 | 0.0038 | 0.0006 | 0.0050 | 0.0044 | 0.0019 | 0.0006 | 0.0006  |\n",
      "|   R    | 0.0000 | 0.0006 | 0.0013 | 0.0006 | 0.0019 | 0.0025 | 0.0019 | 0.0025 | 0.0031 | 0.0038 | 0.0006 | 0.0031 | 0.0006 | 0.0038 | 0.0025 | 0.0050 | 0.0019 | 0.0013 | 0.0013 | 0.0006  |\n",
      "|   S    | 0.0025 | 0.0019 | 0.0025 | 0.0006 | 0.0044 | 0.0006 | 0.0031 | 0.0031 | 0.0038 | 0.0063 | 0.0013 | 0.0025 | 0.0031 | 0.0031 | 0.0031 | 0.0050 | 0.0094 | 0.0057 | 0.0013 | 0.0038  |\n",
      "|   T    | 0.0006 | 0.0025 | 0.0006 | 0.0013 | 0.0050 | 0.0050 | 0.0013 | 0.0031 | 0.0044 | 0.0063 | 0.0031 | 0.0050 | 0.0000 | 0.0031 | 0.0013 | 0.0057 | 0.0069 | 0.0038 | 0.0025 | 0.0013  |\n",
      "|   V    | 0.0044 | 0.0019 | 0.0019 | 0.0019 | 0.0094 | 0.0019 | 0.0031 | 0.0025 | 0.0025 | 0.0151 | 0.0013 | 0.0006 | 0.0019 | 0.0025 | 0.0025 | 0.0069 | 0.0063 | 0.0063 | 0.0000 | 0.0031  |\n",
      "|   W    | 0.0006 | 0.0031 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0006 | 0.0006 | 0.0013 | 0.0025 | 0.0000 | 0.0013 | 0.0000 | 0.0006 | 0.0006 | 0.0019 | 0.0000 | 0.0019 | 0.0006 | 0.0006  |\n",
      "|   Y    | 0.0000 | 0.0013 | 0.0013 | 0.0000 | 0.0019 | 0.0031 | 0.0019 | 0.0019 | 0.0013 | 0.0044 | 0.0000 | 0.0038 | 0.0000 | 0.0013 | 0.0019 | 0.0025 | 0.0019 | 0.0013 | 0.0006 | 0.0025  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\mamalian1.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "458d71ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.065159 |\n",
      "|     C     | 0.036124 |\n",
      "|     D     | 0.025996 |\n",
      "|     E     | 0.033086 |\n",
      "|     F     | 0.046590 |\n",
      "|     G     | 0.064146 |\n",
      "|     H     | 0.027684 |\n",
      "|     I     | 0.037812 |\n",
      "|     K     | 0.038488 |\n",
      "|     L     | 0.085078 |\n",
      "|     M     | 0.082714 |\n",
      "|     N     | 0.033761 |\n",
      "|     P     | 0.068535 |\n",
      "|     Q     | 0.037475 |\n",
      "|     R     | 0.076975 |\n",
      "|     S     | 0.090142 |\n",
      "|     T     | 0.054693 |\n",
      "|     V     | 0.049966 |\n",
      "|     W     | 0.022957 |\n",
      "|     Y     | 0.022620 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0040 | 0.0024 | 0.0016 | 0.0044 | 0.0048 | 0.0048 | 0.0008 | 0.0028 | 0.0048 | 0.0048 | 0.0016 | 0.0028 | 0.0040 | 0.0012 | 0.0016 | 0.0067 | 0.0067 | 0.0032 | 0.0012 | 0.0016  |\n",
      "|   C    | 0.0020 | 0.0012 | 0.0004 | 0.0016 | 0.0020 | 0.0012 | 0.0008 | 0.0036 | 0.0008 | 0.0063 | 0.0008 | 0.0028 | 0.0032 | 0.0004 | 0.0024 | 0.0036 | 0.0004 | 0.0028 | 0.0008 | 0.0012  |\n",
      "|   D    | 0.0036 | 0.0016 | 0.0004 | 0.0004 | 0.0008 | 0.0008 | 0.0012 | 0.0008 | 0.0004 | 0.0024 | 0.0008 | 0.0008 | 0.0028 | 0.0000 | 0.0020 | 0.0016 | 0.0012 | 0.0016 | 0.0008 | 0.0012  |\n",
      "|   E    | 0.0020 | 0.0024 | 0.0020 | 0.0044 | 0.0008 | 0.0008 | 0.0004 | 0.0004 | 0.0004 | 0.0008 | 0.0000 | 0.0012 | 0.0036 | 0.0024 | 0.0040 | 0.0044 | 0.0016 | 0.0012 | 0.0004 | 0.0008  |\n",
      "|   F    | 0.0024 | 0.0004 | 0.0032 | 0.0012 | 0.0016 | 0.0020 | 0.0028 | 0.0024 | 0.0012 | 0.0055 | 0.0008 | 0.0008 | 0.0032 | 0.0059 | 0.0048 | 0.0040 | 0.0020 | 0.0012 | 0.0004 | 0.0016  |\n",
      "|   G    | 0.0075 | 0.0020 | 0.0036 | 0.0024 | 0.0055 | 0.0079 | 0.0012 | 0.0016 | 0.0032 | 0.0040 | 0.0008 | 0.0000 | 0.0048 | 0.0012 | 0.0051 | 0.0055 | 0.0020 | 0.0032 | 0.0012 | 0.0012  |\n",
      "|   H    | 0.0020 | 0.0032 | 0.0004 | 0.0008 | 0.0012 | 0.0028 | 0.0012 | 0.0012 | 0.0004 | 0.0032 | 0.0004 | 0.0012 | 0.0016 | 0.0016 | 0.0028 | 0.0028 | 0.0008 | 0.0012 | 0.0012 | 0.0008  |\n",
      "|   I    | 0.0020 | 0.0032 | 0.0000 | 0.0004 | 0.0040 | 0.0016 | 0.0012 | 0.0032 | 0.0020 | 0.0055 | 0.0000 | 0.0004 | 0.0028 | 0.0016 | 0.0008 | 0.0032 | 0.0016 | 0.0012 | 0.0016 | 0.0004  |\n",
      "|   K    | 0.0032 | 0.0008 | 0.0008 | 0.0016 | 0.0028 | 0.0012 | 0.0012 | 0.0024 | 0.0028 | 0.0048 | 0.0016 | 0.0016 | 0.0016 | 0.0024 | 0.0016 | 0.0024 | 0.0040 | 0.0024 | 0.0004 | 0.0012  |\n",
      "|   L    | 0.0024 | 0.0063 | 0.0024 | 0.0040 | 0.0036 | 0.0024 | 0.0020 | 0.0016 | 0.0016 | 0.0127 | 0.0008 | 0.0008 | 0.0119 | 0.0040 | 0.0059 | 0.0055 | 0.0024 | 0.0044 | 0.0032 | 0.0020  |\n",
      "|   M    | 0.0048 | 0.0020 | 0.0012 | 0.0032 | 0.0020 | 0.0075 | 0.0016 | 0.0032 | 0.0044 | 0.0075 | 0.0028 | 0.0028 | 0.0091 | 0.0036 | 0.0067 | 0.0095 | 0.0020 | 0.0036 | 0.0008 | 0.0012  |\n",
      "|   N    | 0.0024 | 0.0016 | 0.0016 | 0.0000 | 0.0024 | 0.0012 | 0.0016 | 0.0012 | 0.0008 | 0.0040 | 0.0008 | 0.0024 | 0.0032 | 0.0004 | 0.0028 | 0.0016 | 0.0024 | 0.0012 | 0.0008 | 0.0004  |\n",
      "|   P    | 0.0055 | 0.0016 | 0.0040 | 0.0024 | 0.0028 | 0.0040 | 0.0040 | 0.0032 | 0.0020 | 0.0044 | 0.0012 | 0.0032 | 0.0044 | 0.0024 | 0.0075 | 0.0095 | 0.0059 | 0.0040 | 0.0008 | 0.0004  |\n",
      "|   Q    | 0.0028 | 0.0008 | 0.0004 | 0.0008 | 0.0008 | 0.0028 | 0.0012 | 0.0020 | 0.0020 | 0.0020 | 0.0016 | 0.0012 | 0.0020 | 0.0016 | 0.0040 | 0.0020 | 0.0063 | 0.0028 | 0.0000 | 0.0008  |\n",
      "|   R    | 0.0075 | 0.0016 | 0.0020 | 0.0012 | 0.0032 | 0.0051 | 0.0024 | 0.0020 | 0.0036 | 0.0044 | 0.0004 | 0.0028 | 0.0036 | 0.0012 | 0.0099 | 0.0127 | 0.0083 | 0.0048 | 0.0008 | 0.0012  |\n",
      "|   S    | 0.0059 | 0.0016 | 0.0004 | 0.0024 | 0.0048 | 0.0075 | 0.0020 | 0.0024 | 0.0028 | 0.0067 | 0.0012 | 0.0036 | 0.0071 | 0.0063 | 0.0075 | 0.0115 | 0.0059 | 0.0036 | 0.0040 | 0.0020  |\n",
      "|   T    | 0.0036 | 0.0020 | 0.0016 | 0.0020 | 0.0016 | 0.0051 | 0.0020 | 0.0024 | 0.0020 | 0.0036 | 0.0016 | 0.0020 | 0.0036 | 0.0004 | 0.0048 | 0.0040 | 0.0044 | 0.0044 | 0.0036 | 0.0008  |\n",
      "|   V    | 0.0032 | 0.0004 | 0.0008 | 0.0016 | 0.0028 | 0.0059 | 0.0012 | 0.0028 | 0.0012 | 0.0055 | 0.0004 | 0.0008 | 0.0020 | 0.0016 | 0.0044 | 0.0051 | 0.0012 | 0.0036 | 0.0016 | 0.0024  |\n",
      "|   W    | 0.0000 | 0.0016 | 0.0012 | 0.0020 | 0.0004 | 0.0012 | 0.0008 | 0.0008 | 0.0024 | 0.0016 | 0.0024 | 0.0016 | 0.0004 | 0.0000 | 0.0024 | 0.0008 | 0.0020 | 0.0016 | 0.0000 | 0.0012  |\n",
      "|   Y    | 0.0020 | 0.0008 | 0.0008 | 0.0004 | 0.0020 | 0.0004 | 0.0008 | 0.0000 | 0.0020 | 0.0020 | 0.0000 | 0.0016 | 0.0016 | 0.0008 | 0.0024 | 0.0028 | 0.0008 | 0.0008 | 0.0000 | 0.0016  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\mamalian2.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "48194e18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.018357 |\n",
      "|     C     | 0.030157 |\n",
      "|     D     | 0.029283 |\n",
      "|     E     | 0.015297 |\n",
      "|     F     | 0.067745 |\n",
      "|     G     | 0.025787 |\n",
      "|     H     | 0.037150 |\n",
      "|     I     | 0.111014 |\n",
      "|     K     | 0.040647 |\n",
      "|     L     | 0.102710 |\n",
      "|     M     | 0.102273 |\n",
      "|     N     | 0.035839 |\n",
      "|     P     | 0.032343 |\n",
      "|     Q     | 0.024913 |\n",
      "|     R     | 0.054633 |\n",
      "|     S     | 0.086976 |\n",
      "|     T     | 0.049825 |\n",
      "|     V     | 0.057255 |\n",
      "|     W     | 0.013986 |\n",
      "|     Y     | 0.063811 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0000 | 0.0011 | 0.0000 | 0.0017 | 0.0017 | 0.0000 | 0.0006 | 0.0028 | 0.0017 | 0.0006 | 0.0000 | 0.0006 | 0.0006 | 0.0000 | 0.0011 | 0.0022 | 0.0011 | 0.0022 | 0.0000 | 0.0000  |\n",
      "|   C    | 0.0017 | 0.0006 | 0.0011 | 0.0000 | 0.0034 | 0.0017 | 0.0022 | 0.0050 | 0.0011 | 0.0039 | 0.0006 | 0.0028 | 0.0000 | 0.0017 | 0.0011 | 0.0017 | 0.0017 | 0.0022 | 0.0006 | 0.0006  |\n",
      "|   D    | 0.0000 | 0.0006 | 0.0006 | 0.0011 | 0.0017 | 0.0017 | 0.0006 | 0.0056 | 0.0034 | 0.0017 | 0.0011 | 0.0000 | 0.0011 | 0.0000 | 0.0039 | 0.0028 | 0.0011 | 0.0028 | 0.0006 | 0.0011  |\n",
      "|   E    | 0.0000 | 0.0011 | 0.0006 | 0.0000 | 0.0000 | 0.0000 | 0.0006 | 0.0022 | 0.0006 | 0.0011 | 0.0000 | 0.0000 | 0.0006 | 0.0006 | 0.0017 | 0.0022 | 0.0011 | 0.0000 | 0.0000 | 0.0011  |\n",
      "|   F    | 0.0000 | 0.0011 | 0.0022 | 0.0006 | 0.0078 | 0.0028 | 0.0034 | 0.0067 | 0.0000 | 0.0089 | 0.0017 | 0.0017 | 0.0028 | 0.0017 | 0.0056 | 0.0056 | 0.0045 | 0.0039 | 0.0017 | 0.0050  |\n",
      "|   G    | 0.0011 | 0.0022 | 0.0000 | 0.0006 | 0.0022 | 0.0011 | 0.0006 | 0.0039 | 0.0017 | 0.0034 | 0.0006 | 0.0006 | 0.0006 | 0.0000 | 0.0000 | 0.0028 | 0.0022 | 0.0000 | 0.0000 | 0.0017  |\n",
      "|   H    | 0.0011 | 0.0022 | 0.0011 | 0.0000 | 0.0017 | 0.0011 | 0.0050 | 0.0050 | 0.0006 | 0.0045 | 0.0000 | 0.0011 | 0.0022 | 0.0011 | 0.0017 | 0.0006 | 0.0017 | 0.0017 | 0.0006 | 0.0034  |\n",
      "|   I    | 0.0034 | 0.0050 | 0.0034 | 0.0011 | 0.0073 | 0.0011 | 0.0039 | 0.0168 | 0.0022 | 0.0134 | 0.0011 | 0.0022 | 0.0017 | 0.0045 | 0.0084 | 0.0084 | 0.0061 | 0.0061 | 0.0022 | 0.0067  |\n",
      "|   K    | 0.0006 | 0.0011 | 0.0017 | 0.0017 | 0.0045 | 0.0000 | 0.0011 | 0.0056 | 0.0050 | 0.0034 | 0.0000 | 0.0034 | 0.0006 | 0.0006 | 0.0022 | 0.0039 | 0.0000 | 0.0006 | 0.0006 | 0.0022  |\n",
      "|   L    | 0.0011 | 0.0045 | 0.0022 | 0.0056 | 0.0112 | 0.0034 | 0.0045 | 0.0067 | 0.0022 | 0.0084 | 0.0011 | 0.0017 | 0.0050 | 0.0034 | 0.0022 | 0.0128 | 0.0034 | 0.0045 | 0.0034 | 0.0117  |\n",
      "|   M    | 0.0045 | 0.0028 | 0.0045 | 0.0006 | 0.0073 | 0.0022 | 0.0017 | 0.0123 | 0.0067 | 0.0162 | 0.0011 | 0.0056 | 0.0022 | 0.0011 | 0.0067 | 0.0173 | 0.0067 | 0.0095 | 0.0006 | 0.0056  |\n",
      "|   N    | 0.0017 | 0.0000 | 0.0022 | 0.0011 | 0.0017 | 0.0028 | 0.0011 | 0.0045 | 0.0028 | 0.0045 | 0.0000 | 0.0006 | 0.0022 | 0.0006 | 0.0006 | 0.0022 | 0.0028 | 0.0045 | 0.0006 | 0.0017  |\n",
      "|   P    | 0.0011 | 0.0006 | 0.0017 | 0.0006 | 0.0028 | 0.0000 | 0.0000 | 0.0045 | 0.0011 | 0.0045 | 0.0000 | 0.0006 | 0.0028 | 0.0017 | 0.0022 | 0.0017 | 0.0045 | 0.0000 | 0.0000 | 0.0006  |\n",
      "|   Q    | 0.0000 | 0.0006 | 0.0000 | 0.0000 | 0.0045 | 0.0006 | 0.0039 | 0.0045 | 0.0017 | 0.0017 | 0.0000 | 0.0011 | 0.0006 | 0.0000 | 0.0022 | 0.0011 | 0.0017 | 0.0000 | 0.0006 | 0.0017  |\n",
      "|   R    | 0.0017 | 0.0006 | 0.0011 | 0.0000 | 0.0022 | 0.0034 | 0.0022 | 0.0045 | 0.0045 | 0.0067 | 0.0011 | 0.0028 | 0.0011 | 0.0011 | 0.0056 | 0.0050 | 0.0017 | 0.0067 | 0.0000 | 0.0050  |\n",
      "|   S    | 0.0022 | 0.0022 | 0.0017 | 0.0006 | 0.0061 | 0.0017 | 0.0028 | 0.0078 | 0.0056 | 0.0123 | 0.0039 | 0.0022 | 0.0022 | 0.0006 | 0.0073 | 0.0112 | 0.0050 | 0.0073 | 0.0017 | 0.0061  |\n",
      "|   T    | 0.0011 | 0.0006 | 0.0011 | 0.0006 | 0.0000 | 0.0011 | 0.0017 | 0.0056 | 0.0022 | 0.0034 | 0.0011 | 0.0034 | 0.0039 | 0.0022 | 0.0017 | 0.0050 | 0.0028 | 0.0034 | 0.0011 | 0.0034  |\n",
      "|   V    | 0.0000 | 0.0034 | 0.0022 | 0.0011 | 0.0028 | 0.0017 | 0.0011 | 0.0050 | 0.0011 | 0.0067 | 0.0000 | 0.0017 | 0.0034 | 0.0028 | 0.0017 | 0.0067 | 0.0034 | 0.0039 | 0.0006 | 0.0045  |\n",
      "|   W    | 0.0000 | 0.0006 | 0.0006 | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0022 | 0.0006 | 0.0011 | 0.0000 | 0.0011 | 0.0000 | 0.0006 | 0.0017 | 0.0006 | 0.0011 | 0.0000 | 0.0006 | 0.0017  |\n",
      "|   Y    | 0.0000 | 0.0028 | 0.0017 | 0.0017 | 0.0034 | 0.0017 | 0.0034 | 0.0095 | 0.0017 | 0.0067 | 0.0022 | 0.0011 | 0.0050 | 0.0028 | 0.0022 | 0.0061 | 0.0022 | 0.0045 | 0.0006 | 0.0039  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\mamalian3.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "9d874f5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.074598 |\n",
      "|     C     | 0.035592 |\n",
      "|     D     | 0.019015 |\n",
      "|     E     | 0.030717 |\n",
      "|     F     | 0.032179 |\n",
      "|     G     | 0.072160 |\n",
      "|     H     | 0.025353 |\n",
      "|     I     | 0.035592 |\n",
      "|     K     | 0.024866 |\n",
      "|     L     | 0.078986 |\n",
      "|     M     | 0.074110 |\n",
      "|     N     | 0.024378 |\n",
      "|     P     | 0.053145 |\n",
      "|     Q     | 0.031692 |\n",
      "|     R     | 0.103364 |\n",
      "|     S     | 0.108727 |\n",
      "|     T     | 0.059971 |\n",
      "|     V     | 0.072160 |\n",
      "|     W     | 0.023403 |\n",
      "|     Y     | 0.019990 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0128 | 0.0033 | 0.0022 | 0.0011 | 0.0011 | 0.0050 | 0.0011 | 0.0017 | 0.0022 | 0.0028 | 0.0006 | 0.0011 | 0.0044 | 0.0022 | 0.0083 | 0.0044 | 0.0044 | 0.0089 | 0.0044 | 0.0017  |\n",
      "|   C    | 0.0011 | 0.0022 | 0.0000 | 0.0022 | 0.0011 | 0.0028 | 0.0011 | 0.0017 | 0.0000 | 0.0050 | 0.0000 | 0.0006 | 0.0017 | 0.0017 | 0.0050 | 0.0044 | 0.0006 | 0.0033 | 0.0011 | 0.0011  |\n",
      "|   D    | 0.0028 | 0.0000 | 0.0006 | 0.0000 | 0.0000 | 0.0017 | 0.0006 | 0.0006 | 0.0000 | 0.0022 | 0.0006 | 0.0000 | 0.0006 | 0.0011 | 0.0028 | 0.0011 | 0.0006 | 0.0028 | 0.0006 | 0.0006  |\n",
      "|   E    | 0.0011 | 0.0028 | 0.0000 | 0.0039 | 0.0006 | 0.0017 | 0.0006 | 0.0017 | 0.0006 | 0.0017 | 0.0022 | 0.0017 | 0.0011 | 0.0000 | 0.0022 | 0.0033 | 0.0022 | 0.0022 | 0.0000 | 0.0017  |\n",
      "|   F    | 0.0006 | 0.0017 | 0.0000 | 0.0006 | 0.0028 | 0.0028 | 0.0006 | 0.0000 | 0.0000 | 0.0022 | 0.0006 | 0.0006 | 0.0033 | 0.0017 | 0.0022 | 0.0044 | 0.0044 | 0.0017 | 0.0011 | 0.0000  |\n",
      "|   G    | 0.0044 | 0.0028 | 0.0050 | 0.0039 | 0.0017 | 0.0089 | 0.0044 | 0.0011 | 0.0022 | 0.0022 | 0.0006 | 0.0033 | 0.0033 | 0.0017 | 0.0067 | 0.0100 | 0.0028 | 0.0067 | 0.0011 | 0.0000  |\n",
      "|   H    | 0.0022 | 0.0000 | 0.0017 | 0.0000 | 0.0011 | 0.0011 | 0.0006 | 0.0006 | 0.0006 | 0.0022 | 0.0006 | 0.0000 | 0.0017 | 0.0017 | 0.0055 | 0.0017 | 0.0006 | 0.0017 | 0.0000 | 0.0022  |\n",
      "|   I    | 0.0011 | 0.0022 | 0.0006 | 0.0022 | 0.0011 | 0.0028 | 0.0006 | 0.0011 | 0.0006 | 0.0055 | 0.0006 | 0.0000 | 0.0006 | 0.0011 | 0.0033 | 0.0033 | 0.0039 | 0.0028 | 0.0011 | 0.0000  |\n",
      "|   K    | 0.0006 | 0.0000 | 0.0000 | 0.0022 | 0.0006 | 0.0011 | 0.0017 | 0.0011 | 0.0022 | 0.0017 | 0.0000 | 0.0017 | 0.0011 | 0.0011 | 0.0033 | 0.0022 | 0.0033 | 0.0006 | 0.0006 | 0.0000  |\n",
      "|   L    | 0.0061 | 0.0011 | 0.0011 | 0.0011 | 0.0050 | 0.0039 | 0.0028 | 0.0011 | 0.0006 | 0.0111 | 0.0000 | 0.0000 | 0.0050 | 0.0033 | 0.0100 | 0.0067 | 0.0011 | 0.0083 | 0.0033 | 0.0022  |\n",
      "|   M    | 0.0061 | 0.0028 | 0.0017 | 0.0033 | 0.0033 | 0.0055 | 0.0028 | 0.0044 | 0.0017 | 0.0067 | 0.0017 | 0.0033 | 0.0011 | 0.0000 | 0.0061 | 0.0072 | 0.0050 | 0.0033 | 0.0000 | 0.0033  |\n",
      "|   N    | 0.0017 | 0.0011 | 0.0011 | 0.0006 | 0.0000 | 0.0022 | 0.0011 | 0.0006 | 0.0011 | 0.0011 | 0.0006 | 0.0000 | 0.0011 | 0.0006 | 0.0028 | 0.0033 | 0.0017 | 0.0028 | 0.0000 | 0.0000  |\n",
      "|   P    | 0.0044 | 0.0028 | 0.0000 | 0.0017 | 0.0033 | 0.0028 | 0.0006 | 0.0011 | 0.0000 | 0.0039 | 0.0006 | 0.0011 | 0.0044 | 0.0006 | 0.0061 | 0.0067 | 0.0044 | 0.0067 | 0.0022 | 0.0017  |\n",
      "|   Q    | 0.0022 | 0.0000 | 0.0000 | 0.0006 | 0.0011 | 0.0022 | 0.0017 | 0.0022 | 0.0017 | 0.0044 | 0.0006 | 0.0011 | 0.0017 | 0.0028 | 0.0055 | 0.0022 | 0.0006 | 0.0017 | 0.0006 | 0.0000  |\n",
      "|   R    | 0.0083 | 0.0055 | 0.0000 | 0.0017 | 0.0006 | 0.0117 | 0.0039 | 0.0050 | 0.0028 | 0.0100 | 0.0006 | 0.0022 | 0.0067 | 0.0039 | 0.0144 | 0.0122 | 0.0072 | 0.0055 | 0.0017 | 0.0028  |\n",
      "|   S    | 0.0089 | 0.0022 | 0.0022 | 0.0028 | 0.0028 | 0.0111 | 0.0000 | 0.0033 | 0.0039 | 0.0072 | 0.0022 | 0.0055 | 0.0100 | 0.0039 | 0.0122 | 0.0216 | 0.0055 | 0.0061 | 0.0028 | 0.0011  |\n",
      "|   T    | 0.0083 | 0.0028 | 0.0000 | 0.0017 | 0.0022 | 0.0011 | 0.0017 | 0.0017 | 0.0028 | 0.0055 | 0.0006 | 0.0017 | 0.0028 | 0.0011 | 0.0055 | 0.0105 | 0.0050 | 0.0044 | 0.0000 | 0.0006  |\n",
      "|   V    | 0.0033 | 0.0028 | 0.0022 | 0.0028 | 0.0039 | 0.0083 | 0.0006 | 0.0028 | 0.0022 | 0.0067 | 0.0011 | 0.0011 | 0.0055 | 0.0011 | 0.0022 | 0.0122 | 0.0044 | 0.0055 | 0.0011 | 0.0006  |\n",
      "|   W    | 0.0017 | 0.0006 | 0.0000 | 0.0006 | 0.0000 | 0.0006 | 0.0000 | 0.0017 | 0.0006 | 0.0017 | 0.0000 | 0.0006 | 0.0011 | 0.0011 | 0.0039 | 0.0017 | 0.0022 | 0.0028 | 0.0028 | 0.0011  |\n",
      "|   Y    | 0.0017 | 0.0011 | 0.0000 | 0.0006 | 0.0028 | 0.0011 | 0.0000 | 0.0017 | 0.0006 | 0.0017 | 0.0000 | 0.0006 | 0.0006 | 0.0011 | 0.0011 | 0.0006 | 0.0028 | 0.0000 | 0.0017 | 0.0006  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\mamalian4.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "751b5032",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.039463 |\n",
      "|     C     | 0.029203 |\n",
      "|     D     | 0.018942 |\n",
      "|     E     | 0.026835 |\n",
      "|     F     | 0.061563 |\n",
      "|     G     | 0.037885 |\n",
      "|     H     | 0.022099 |\n",
      "|     I     | 0.064720 |\n",
      "|     K     | 0.062352 |\n",
      "|     L     | 0.092344 |\n",
      "|     M     | 0.138122 |\n",
      "|     N     | 0.037096 |\n",
      "|     P     | 0.021310 |\n",
      "|     Q     | 0.051302 |\n",
      "|     R     | 0.050513 |\n",
      "|     S     | 0.089187 |\n",
      "|     T     | 0.049724 |\n",
      "|     V     | 0.050513 |\n",
      "|     W     | 0.012628 |\n",
      "|     Y     | 0.044199 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0010 | 0.0042 | 0.0000 | 0.0042 | 0.0000 | 0.0010 | 0.0000 | 0.0031 | 0.0042 | 0.0063 | 0.0010 | 0.0042 | 0.0010 | 0.0000 | 0.0021 | 0.0010 | 0.0021 | 0.0010 | 0.0000 | 0.0031  |\n",
      "|   C    | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0042 | 0.0010 | 0.0000 | 0.0010 | 0.0021 | 0.0000 | 0.0000 | 0.0010 | 0.0000 | 0.0021 | 0.0000 | 0.0052 | 0.0000 | 0.0010 | 0.0000 | 0.0042  |\n",
      "|   D    | 0.0010 | 0.0000 | 0.0000 | 0.0000 | 0.0031 | 0.0010 | 0.0010 | 0.0021 | 0.0010 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0042 | 0.0021 | 0.0000 | 0.0000 | 0.0021 | 0.0000  |\n",
      "|   E    | 0.0010 | 0.0000 | 0.0000 | 0.0000 | 0.0010 | 0.0010 | 0.0000 | 0.0010 | 0.0010 | 0.0010 | 0.0021 | 0.0021 | 0.0000 | 0.0042 | 0.0052 | 0.0021 | 0.0010 | 0.0010 | 0.0010 | 0.0010  |\n",
      "|   F    | 0.0031 | 0.0052 | 0.0010 | 0.0021 | 0.0063 | 0.0021 | 0.0021 | 0.0084 | 0.0052 | 0.0105 | 0.0010 | 0.0021 | 0.0010 | 0.0000 | 0.0042 | 0.0073 | 0.0010 | 0.0021 | 0.0000 | 0.0021  |\n",
      "|   G    | 0.0010 | 0.0000 | 0.0010 | 0.0000 | 0.0042 | 0.0031 | 0.0010 | 0.0084 | 0.0042 | 0.0052 | 0.0000 | 0.0010 | 0.0000 | 0.0021 | 0.0021 | 0.0042 | 0.0031 | 0.0010 | 0.0000 | 0.0000  |\n",
      "|   H    | 0.0031 | 0.0010 | 0.0000 | 0.0010 | 0.0021 | 0.0010 | 0.0000 | 0.0000 | 0.0031 | 0.0010 | 0.0000 | 0.0000 | 0.0000 | 0.0010 | 0.0010 | 0.0021 | 0.0010 | 0.0021 | 0.0000 | 0.0000  |\n",
      "|   I    | 0.0031 | 0.0010 | 0.0010 | 0.0000 | 0.0042 | 0.0021 | 0.0021 | 0.0021 | 0.0010 | 0.0084 | 0.0000 | 0.0021 | 0.0010 | 0.0073 | 0.0052 | 0.0052 | 0.0010 | 0.0021 | 0.0000 | 0.0052  |\n",
      "|   K    | 0.0021 | 0.0031 | 0.0000 | 0.0010 | 0.0052 | 0.0021 | 0.0021 | 0.0042 | 0.0042 | 0.0073 | 0.0010 | 0.0021 | 0.0031 | 0.0021 | 0.0063 | 0.0010 | 0.0094 | 0.0052 | 0.0010 | 0.0021  |\n",
      "|   L    | 0.0031 | 0.0010 | 0.0031 | 0.0021 | 0.0042 | 0.0031 | 0.0052 | 0.0063 | 0.0021 | 0.0115 | 0.0021 | 0.0010 | 0.0021 | 0.0115 | 0.0021 | 0.0115 | 0.0031 | 0.0042 | 0.0021 | 0.0094  |\n",
      "|   M    | 0.0136 | 0.0021 | 0.0052 | 0.0115 | 0.0105 | 0.0063 | 0.0021 | 0.0052 | 0.0115 | 0.0105 | 0.0000 | 0.0073 | 0.0063 | 0.0084 | 0.0010 | 0.0073 | 0.0136 | 0.0115 | 0.0010 | 0.0094  |\n",
      "|   N    | 0.0031 | 0.0021 | 0.0010 | 0.0000 | 0.0063 | 0.0052 | 0.0010 | 0.0031 | 0.0063 | 0.0031 | 0.0000 | 0.0000 | 0.0010 | 0.0000 | 0.0000 | 0.0021 | 0.0010 | 0.0021 | 0.0010 | 0.0000  |\n",
      "|   P    | 0.0021 | 0.0010 | 0.0000 | 0.0000 | 0.0021 | 0.0010 | 0.0010 | 0.0031 | 0.0000 | 0.0010 | 0.0010 | 0.0000 | 0.0010 | 0.0000 | 0.0021 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0010  |\n",
      "|   Q    | 0.0010 | 0.0000 | 0.0000 | 0.0000 | 0.0010 | 0.0000 | 0.0000 | 0.0042 | 0.0052 | 0.0063 | 0.0010 | 0.0021 | 0.0010 | 0.0021 | 0.0031 | 0.0031 | 0.0063 | 0.0073 | 0.0010 | 0.0021  |\n",
      "|   R    | 0.0021 | 0.0031 | 0.0021 | 0.0010 | 0.0042 | 0.0010 | 0.0010 | 0.0021 | 0.0021 | 0.0031 | 0.0000 | 0.0021 | 0.0010 | 0.0042 | 0.0031 | 0.0094 | 0.0042 | 0.0042 | 0.0010 | 0.0000  |\n",
      "|   S    | 0.0010 | 0.0052 | 0.0010 | 0.0021 | 0.0042 | 0.0063 | 0.0021 | 0.0105 | 0.0063 | 0.0115 | 0.0031 | 0.0063 | 0.0021 | 0.0021 | 0.0021 | 0.0168 | 0.0021 | 0.0063 | 0.0010 | 0.0052  |\n",
      "|   T    | 0.0010 | 0.0031 | 0.0000 | 0.0031 | 0.0000 | 0.0021 | 0.0000 | 0.0021 | 0.0052 | 0.0115 | 0.0021 | 0.0021 | 0.0010 | 0.0042 | 0.0031 | 0.0010 | 0.0021 | 0.0042 | 0.0021 | 0.0021  |\n",
      "|   V    | 0.0010 | 0.0010 | 0.0031 | 0.0021 | 0.0052 | 0.0031 | 0.0010 | 0.0031 | 0.0042 | 0.0094 | 0.0000 | 0.0010 | 0.0010 | 0.0042 | 0.0042 | 0.0031 | 0.0031 | 0.0010 | 0.0010 | 0.0021  |\n",
      "|   W    | 0.0000 | 0.0000 | 0.0010 | 0.0010 | 0.0010 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0010 | 0.0010 | 0.0010 | 0.0000 | 0.0021 | 0.0010 | 0.0031 | 0.0010 | 0.0000 | 0.0000 | 0.0000  |\n",
      "|   Y    | 0.0031 | 0.0021 | 0.0021 | 0.0000 | 0.0031 | 0.0021 | 0.0042 | 0.0042 | 0.0000 | 0.0000 | 0.0000 | 0.0010 | 0.0000 | 0.0031 | 0.0021 | 0.0073 | 0.0000 | 0.0000 | 0.0010 | 0.0021  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\bacterial1.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "d76c8777",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.045950 |\n",
      "|     C     | 0.031542 |\n",
      "|     D     | 0.027259 |\n",
      "|     E     | 0.022586 |\n",
      "|     F     | 0.043614 |\n",
      "|     G     | 0.047118 |\n",
      "|     H     | 0.029595 |\n",
      "|     I     | 0.061137 |\n",
      "|     K     | 0.038551 |\n",
      "|     L     | 0.103583 |\n",
      "|     M     | 0.091121 |\n",
      "|     N     | 0.029984 |\n",
      "|     P     | 0.059579 |\n",
      "|     Q     | 0.035047 |\n",
      "|     R     | 0.069704 |\n",
      "|     S     | 0.090732 |\n",
      "|     T     | 0.047508 |\n",
      "|     V     | 0.072430 |\n",
      "|     W     | 0.020639 |\n",
      "|     Y     | 0.032321 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0018 | 0.0018 | 0.0009 | 0.0014 | 0.0046 | 0.0018 | 0.0014 | 0.0018 | 0.0036 | 0.0036 | 0.0018 | 0.0014 | 0.0023 | 0.0018 | 0.0027 | 0.0050 | 0.0027 | 0.0032 | 0.0005 | 0.0009  |\n",
      "|   C    | 0.0005 | 0.0018 | 0.0005 | 0.0000 | 0.0023 | 0.0018 | 0.0005 | 0.0046 | 0.0005 | 0.0041 | 0.0000 | 0.0009 | 0.0018 | 0.0014 | 0.0023 | 0.0041 | 0.0000 | 0.0036 | 0.0009 | 0.0014  |\n",
      "|   D    | 0.0005 | 0.0009 | 0.0000 | 0.0009 | 0.0032 | 0.0009 | 0.0005 | 0.0027 | 0.0009 | 0.0027 | 0.0005 | 0.0014 | 0.0000 | 0.0005 | 0.0023 | 0.0046 | 0.0018 | 0.0032 | 0.0000 | 0.0009  |\n",
      "|   E    | 0.0000 | 0.0005 | 0.0005 | 0.0000 | 0.0009 | 0.0009 | 0.0005 | 0.0018 | 0.0009 | 0.0018 | 0.0005 | 0.0014 | 0.0027 | 0.0009 | 0.0018 | 0.0014 | 0.0014 | 0.0005 | 0.0000 | 0.0014  |\n",
      "|   F    | 0.0032 | 0.0023 | 0.0005 | 0.0014 | 0.0023 | 0.0032 | 0.0005 | 0.0023 | 0.0018 | 0.0073 | 0.0018 | 0.0005 | 0.0027 | 0.0023 | 0.0027 | 0.0055 | 0.0014 | 0.0018 | 0.0014 | 0.0014  |\n",
      "|   G    | 0.0027 | 0.0014 | 0.0023 | 0.0014 | 0.0059 | 0.0032 | 0.0014 | 0.0041 | 0.0009 | 0.0032 | 0.0018 | 0.0009 | 0.0005 | 0.0009 | 0.0046 | 0.0018 | 0.0018 | 0.0059 | 0.0009 | 0.0041  |\n",
      "|   H    | 0.0014 | 0.0009 | 0.0005 | 0.0009 | 0.0023 | 0.0027 | 0.0023 | 0.0014 | 0.0005 | 0.0009 | 0.0023 | 0.0005 | 0.0036 | 0.0014 | 0.0032 | 0.0041 | 0.0005 | 0.0018 | 0.0005 | 0.0009  |\n",
      "|   I    | 0.0023 | 0.0009 | 0.0000 | 0.0009 | 0.0018 | 0.0018 | 0.0009 | 0.0046 | 0.0014 | 0.0141 | 0.0009 | 0.0009 | 0.0041 | 0.0027 | 0.0041 | 0.0055 | 0.0036 | 0.0036 | 0.0005 | 0.0023  |\n",
      "|   K    | 0.0009 | 0.0018 | 0.0018 | 0.0000 | 0.0009 | 0.0018 | 0.0000 | 0.0023 | 0.0014 | 0.0046 | 0.0009 | 0.0023 | 0.0027 | 0.0009 | 0.0073 | 0.0055 | 0.0023 | 0.0009 | 0.0005 | 0.0005  |\n",
      "|   L    | 0.0032 | 0.0050 | 0.0027 | 0.0009 | 0.0046 | 0.0059 | 0.0036 | 0.0027 | 0.0009 | 0.0128 | 0.0005 | 0.0009 | 0.0091 | 0.0032 | 0.0087 | 0.0114 | 0.0046 | 0.0114 | 0.0009 | 0.0027  |\n",
      "|   M    | 0.0050 | 0.0009 | 0.0032 | 0.0050 | 0.0041 | 0.0036 | 0.0027 | 0.0082 | 0.0064 | 0.0128 | 0.0014 | 0.0064 | 0.0036 | 0.0036 | 0.0059 | 0.0068 | 0.0064 | 0.0077 | 0.0023 | 0.0027  |\n",
      "|   N    | 0.0005 | 0.0023 | 0.0027 | 0.0014 | 0.0018 | 0.0023 | 0.0005 | 0.0023 | 0.0014 | 0.0036 | 0.0005 | 0.0005 | 0.0005 | 0.0000 | 0.0036 | 0.0032 | 0.0023 | 0.0009 | 0.0005 | 0.0014  |\n",
      "|   P    | 0.0059 | 0.0014 | 0.0023 | 0.0018 | 0.0036 | 0.0014 | 0.0014 | 0.0023 | 0.0023 | 0.0073 | 0.0005 | 0.0018 | 0.0046 | 0.0014 | 0.0023 | 0.0077 | 0.0046 | 0.0018 | 0.0023 | 0.0036  |\n",
      "|   Q    | 0.0005 | 0.0014 | 0.0018 | 0.0014 | 0.0005 | 0.0014 | 0.0023 | 0.0050 | 0.0005 | 0.0018 | 0.0014 | 0.0014 | 0.0009 | 0.0023 | 0.0027 | 0.0023 | 0.0009 | 0.0036 | 0.0009 | 0.0018  |\n",
      "|   R    | 0.0014 | 0.0032 | 0.0023 | 0.0018 | 0.0005 | 0.0046 | 0.0041 | 0.0023 | 0.0032 | 0.0077 | 0.0009 | 0.0032 | 0.0036 | 0.0036 | 0.0073 | 0.0050 | 0.0023 | 0.0055 | 0.0023 | 0.0027  |\n",
      "|   S    | 0.0041 | 0.0023 | 0.0014 | 0.0018 | 0.0036 | 0.0055 | 0.0018 | 0.0050 | 0.0041 | 0.0100 | 0.0018 | 0.0014 | 0.0082 | 0.0018 | 0.0041 | 0.0118 | 0.0064 | 0.0082 | 0.0046 | 0.0023  |\n",
      "|   T    | 0.0041 | 0.0009 | 0.0005 | 0.0005 | 0.0023 | 0.0027 | 0.0000 | 0.0027 | 0.0055 | 0.0073 | 0.0009 | 0.0014 | 0.0046 | 0.0018 | 0.0018 | 0.0023 | 0.0036 | 0.0023 | 0.0005 | 0.0009  |\n",
      "|   V    | 0.0032 | 0.0046 | 0.0018 | 0.0009 | 0.0023 | 0.0041 | 0.0014 | 0.0032 | 0.0014 | 0.0082 | 0.0014 | 0.0027 | 0.0046 | 0.0032 | 0.0059 | 0.0068 | 0.0023 | 0.0082 | 0.0014 | 0.0023  |\n",
      "|   W    | 0.0018 | 0.0005 | 0.0014 | 0.0000 | 0.0000 | 0.0005 | 0.0014 | 0.0009 | 0.0023 | 0.0023 | 0.0000 | 0.0014 | 0.0014 | 0.0005 | 0.0009 | 0.0023 | 0.0009 | 0.0018 | 0.0009 | 0.0014  |\n",
      "|   Y    | 0.0046 | 0.0009 | 0.0009 | 0.0014 | 0.0009 | 0.0014 | 0.0041 | 0.0027 | 0.0018 | 0.0005 | 0.0005 | 0.0005 | 0.0032 | 0.0027 | 0.0014 | 0.0027 | 0.0009 | 0.0000 | 0.0014 | 0.0005  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\bacterial2.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "22b031a1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.046138 |\n",
      "|     C     | 0.023847 |\n",
      "|     D     | 0.029549 |\n",
      "|     E     | 0.021773 |\n",
      "|     F     | 0.031104 |\n",
      "|     G     | 0.041991 |\n",
      "|     H     | 0.018144 |\n",
      "|     I     | 0.052359 |\n",
      "|     K     | 0.064800 |\n",
      "|     L     | 0.114567 |\n",
      "|     M     | 0.131156 |\n",
      "|     N     | 0.040435 |\n",
      "|     P     | 0.022291 |\n",
      "|     Q     | 0.044583 |\n",
      "|     R     | 0.071021 |\n",
      "|     S     | 0.076724 |\n",
      "|     T     | 0.051322 |\n",
      "|     V     | 0.054432 |\n",
      "|     W     | 0.025920 |\n",
      "|     Y     | 0.037843 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0032 | 0.0000 | 0.0013 | 0.0019 | 0.0013 | 0.0019 | 0.0000 | 0.0019 | 0.0039 | 0.0084 | 0.0006 | 0.0013 | 0.0000 | 0.0039 | 0.0039 | 0.0052 | 0.0013 | 0.0045 | 0.0000 | 0.0013  |\n",
      "|   C    | 0.0006 | 0.0006 | 0.0013 | 0.0013 | 0.0000 | 0.0000 | 0.0013 | 0.0013 | 0.0032 | 0.0026 | 0.0000 | 0.0019 | 0.0006 | 0.0000 | 0.0013 | 0.0013 | 0.0019 | 0.0006 | 0.0019 | 0.0013  |\n",
      "|   D    | 0.0019 | 0.0000 | 0.0032 | 0.0013 | 0.0000 | 0.0019 | 0.0000 | 0.0039 | 0.0013 | 0.0006 | 0.0019 | 0.0006 | 0.0000 | 0.0013 | 0.0026 | 0.0026 | 0.0006 | 0.0006 | 0.0013 | 0.0013  |\n",
      "|   E    | 0.0006 | 0.0000 | 0.0006 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0013 | 0.0026 | 0.0032 | 0.0019 | 0.0019 | 0.0000 | 0.0013 | 0.0026 | 0.0013 | 0.0006 | 0.0045 | 0.0000 | 0.0013  |\n",
      "|   F    | 0.0013 | 0.0006 | 0.0000 | 0.0006 | 0.0006 | 0.0026 | 0.0006 | 0.0013 | 0.0032 | 0.0026 | 0.0006 | 0.0000 | 0.0000 | 0.0032 | 0.0019 | 0.0026 | 0.0032 | 0.0013 | 0.0006 | 0.0026  |\n",
      "|   G    | 0.0039 | 0.0006 | 0.0006 | 0.0000 | 0.0006 | 0.0026 | 0.0000 | 0.0039 | 0.0052 | 0.0058 | 0.0006 | 0.0026 | 0.0006 | 0.0032 | 0.0032 | 0.0019 | 0.0013 | 0.0032 | 0.0000 | 0.0032  |\n",
      "|   H    | 0.0013 | 0.0013 | 0.0000 | 0.0000 | 0.0000 | 0.0026 | 0.0000 | 0.0000 | 0.0019 | 0.0032 | 0.0000 | 0.0006 | 0.0006 | 0.0006 | 0.0013 | 0.0019 | 0.0006 | 0.0006 | 0.0000 | 0.0006  |\n",
      "|   I    | 0.0026 | 0.0026 | 0.0006 | 0.0006 | 0.0019 | 0.0019 | 0.0006 | 0.0052 | 0.0039 | 0.0078 | 0.0006 | 0.0006 | 0.0006 | 0.0019 | 0.0039 | 0.0052 | 0.0071 | 0.0019 | 0.0032 | 0.0019  |\n",
      "|   K    | 0.0000 | 0.0006 | 0.0006 | 0.0013 | 0.0006 | 0.0019 | 0.0013 | 0.0026 | 0.0097 | 0.0071 | 0.0039 | 0.0084 | 0.0013 | 0.0006 | 0.0071 | 0.0039 | 0.0052 | 0.0045 | 0.0013 | 0.0006  |\n",
      "|   L    | 0.0052 | 0.0026 | 0.0039 | 0.0026 | 0.0104 | 0.0039 | 0.0032 | 0.0065 | 0.0052 | 0.0188 | 0.0032 | 0.0026 | 0.0026 | 0.0032 | 0.0052 | 0.0149 | 0.0045 | 0.0084 | 0.0032 | 0.0026  |\n",
      "|   M    | 0.0110 | 0.0013 | 0.0058 | 0.0045 | 0.0058 | 0.0071 | 0.0006 | 0.0078 | 0.0091 | 0.0194 | 0.0026 | 0.0052 | 0.0058 | 0.0084 | 0.0084 | 0.0156 | 0.0123 | 0.0110 | 0.0019 | 0.0078  |\n",
      "|   N    | 0.0000 | 0.0000 | 0.0000 | 0.0019 | 0.0013 | 0.0006 | 0.0026 | 0.0026 | 0.0019 | 0.0078 | 0.0000 | 0.0019 | 0.0013 | 0.0013 | 0.0039 | 0.0019 | 0.0000 | 0.0000 | 0.0006 | 0.0039  |\n",
      "|   P    | 0.0013 | 0.0013 | 0.0006 | 0.0013 | 0.0000 | 0.0000 | 0.0000 | 0.0006 | 0.0006 | 0.0032 | 0.0019 | 0.0006 | 0.0000 | 0.0013 | 0.0019 | 0.0013 | 0.0013 | 0.0000 | 0.0013 | 0.0000  |\n",
      "|   Q    | 0.0039 | 0.0013 | 0.0013 | 0.0000 | 0.0006 | 0.0019 | 0.0013 | 0.0026 | 0.0039 | 0.0039 | 0.0013 | 0.0026 | 0.0032 | 0.0019 | 0.0065 | 0.0013 | 0.0039 | 0.0032 | 0.0006 | 0.0019  |\n",
      "|   R    | 0.0032 | 0.0032 | 0.0052 | 0.0019 | 0.0013 | 0.0026 | 0.0019 | 0.0039 | 0.0078 | 0.0078 | 0.0006 | 0.0039 | 0.0013 | 0.0032 | 0.0045 | 0.0039 | 0.0032 | 0.0032 | 0.0045 | 0.0019  |\n",
      "|   S    | 0.0000 | 0.0006 | 0.0026 | 0.0019 | 0.0058 | 0.0058 | 0.0019 | 0.0071 | 0.0045 | 0.0091 | 0.0026 | 0.0058 | 0.0039 | 0.0045 | 0.0058 | 0.0065 | 0.0026 | 0.0013 | 0.0013 | 0.0013  |\n",
      "|   T    | 0.0032 | 0.0039 | 0.0013 | 0.0000 | 0.0032 | 0.0026 | 0.0006 | 0.0045 | 0.0006 | 0.0052 | 0.0006 | 0.0013 | 0.0000 | 0.0013 | 0.0045 | 0.0045 | 0.0052 | 0.0039 | 0.0006 | 0.0032  |\n",
      "|   V    | 0.0052 | 0.0013 | 0.0032 | 0.0000 | 0.0019 | 0.0032 | 0.0006 | 0.0019 | 0.0013 | 0.0084 | 0.0026 | 0.0013 | 0.0000 | 0.0039 | 0.0019 | 0.0058 | 0.0013 | 0.0039 | 0.0026 | 0.0032  |\n",
      "|   W    | 0.0013 | 0.0006 | 0.0000 | 0.0006 | 0.0000 | 0.0013 | 0.0006 | 0.0006 | 0.0019 | 0.0019 | 0.0006 | 0.0000 | 0.0000 | 0.0000 | 0.0045 | 0.0032 | 0.0019 | 0.0032 | 0.0013 | 0.0006  |\n",
      "|   Y    | 0.0013 | 0.0013 | 0.0000 | 0.0006 | 0.0019 | 0.0032 | 0.0013 | 0.0006 | 0.0019 | 0.0019 | 0.0000 | 0.0013 | 0.0006 | 0.0039 | 0.0052 | 0.0013 | 0.0000 | 0.0026 | 0.0013 | 0.0026  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\bacterial3.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "3ffba35c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------\n",
      "|Amino Acid | Frequency|\n",
      "------------------------\n",
      "|     A     | 0.022508 |\n",
      "|     C     | 0.020900 |\n",
      "|     D     | 0.017685 |\n",
      "|     E     | 0.016881 |\n",
      "|     F     | 0.064309 |\n",
      "|     G     | 0.040997 |\n",
      "|     H     | 0.012058 |\n",
      "|     I     | 0.073151 |\n",
      "|     K     | 0.081190 |\n",
      "|     L     | 0.125402 |\n",
      "|     M     | 0.216238 |\n",
      "|     N     | 0.045820 |\n",
      "|     P     | 0.011254 |\n",
      "|     Q     | 0.035370 |\n",
      "|     R     | 0.049035 |\n",
      "|     S     | 0.040193 |\n",
      "|     T     | 0.023312 |\n",
      "|     V     | 0.047428 |\n",
      "|     W     | 0.012058 |\n",
      "|     Y     | 0.044212 |\n",
      "------------------------\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|        |   A    |   C    |   D    |   E    |   F    |   G    |   H    |   I    |   K    |   L    |   M    |   N    |   P    |   Q    |   R    |   S    |   T    |   V    |   W    |   Y    |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "|   A    | 0.0033 | 0.0000 | 0.0011 | 0.0000 | 0.0011 | 0.0022 | 0.0000 | 0.0011 | 0.0022 | 0.0022 | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0000 | 0.0033 | 0.0022 | 0.0022 | 0.0011 | 0.0011  |\n",
      "|   C    | 0.0011 | 0.0011 | 0.0000 | 0.0000 | 0.0033 | 0.0011 | 0.0000 | 0.0011 | 0.0022 | 0.0022 | 0.0011 | 0.0011 | 0.0000 | 0.0022 | 0.0011 | 0.0022 | 0.0011 | 0.0022 | 0.0000 | 0.0011  |\n",
      "|   D    | 0.0011 | 0.0000 | 0.0011 | 0.0000 | 0.0011 | 0.0011 | 0.0000 | 0.0022 | 0.0000 | 0.0067 | 0.0011 | 0.0000 | 0.0000 | 0.0000 | 0.0011 | 0.0022 | 0.0011 | 0.0000 | 0.0000 | 0.0022  |\n",
      "|   E    | 0.0011 | 0.0000 | 0.0000 | 0.0011 | 0.0011 | 0.0011 | 0.0000 | 0.0011 | 0.0011 | 0.0033 | 0.0000 | 0.0011 | 0.0011 | 0.0022 | 0.0011 | 0.0022 | 0.0000 | 0.0011 | 0.0000 | 0.0011  |\n",
      "|   F    | 0.0022 | 0.0033 | 0.0000 | 0.0000 | 0.0033 | 0.0033 | 0.0011 | 0.0078 | 0.0022 | 0.0089 | 0.0022 | 0.0011 | 0.0011 | 0.0011 | 0.0056 | 0.0011 | 0.0022 | 0.0011 | 0.0011 | 0.0056  |\n",
      "|   G    | 0.0022 | 0.0000 | 0.0000 | 0.0000 | 0.0033 | 0.0011 | 0.0000 | 0.0033 | 0.0067 | 0.0056 | 0.0000 | 0.0033 | 0.0011 | 0.0011 | 0.0011 | 0.0033 | 0.0022 | 0.0033 | 0.0000 | 0.0000  |\n",
      "|   H    | 0.0011 | 0.0000 | 0.0000 | 0.0000 | 0.0022 | 0.0022 | 0.0022 | 0.0011 | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0022 | 0.0000 | 0.0000 | 0.0000 | 0.0022  |\n",
      "|   I    | 0.0011 | 0.0045 | 0.0000 | 0.0000 | 0.0022 | 0.0033 | 0.0011 | 0.0045 | 0.0022 | 0.0212 | 0.0022 | 0.0067 | 0.0011 | 0.0022 | 0.0078 | 0.0045 | 0.0022 | 0.0011 | 0.0000 | 0.0045  |\n",
      "|   K    | 0.0000 | 0.0000 | 0.0011 | 0.0033 | 0.0067 | 0.0045 | 0.0022 | 0.0078 | 0.0167 | 0.0156 | 0.0011 | 0.0045 | 0.0000 | 0.0000 | 0.0089 | 0.0000 | 0.0022 | 0.0056 | 0.0011 | 0.0045  |\n",
      "|   L    | 0.0000 | 0.0033 | 0.0033 | 0.0011 | 0.0111 | 0.0022 | 0.0033 | 0.0056 | 0.0033 | 0.0245 | 0.0033 | 0.0022 | 0.0011 | 0.0045 | 0.0056 | 0.0067 | 0.0011 | 0.0056 | 0.0022 | 0.0022  |\n",
      "|   M    | 0.0045 | 0.0000 | 0.0111 | 0.0067 | 0.0256 | 0.0134 | 0.0022 | 0.0312 | 0.0334 | 0.0312 | 0.0100 | 0.0089 | 0.0033 | 0.0223 | 0.0067 | 0.0078 | 0.0011 | 0.0200 | 0.0022 | 0.0100  |\n",
      "|   N    | 0.0011 | 0.0000 | 0.0000 | 0.0000 | 0.0011 | 0.0022 | 0.0000 | 0.0033 | 0.0022 | 0.0067 | 0.0022 | 0.0033 | 0.0011 | 0.0011 | 0.0011 | 0.0011 | 0.0022 | 0.0033 | 0.0000 | 0.0056  |\n",
      "|   P    | 0.0000 | 0.0000 | 0.0000 | 0.0011 | 0.0000 | 0.0022 | 0.0000 | 0.0000 | 0.0000 | 0.0011 | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0022  |\n",
      "|   Q    | 0.0033 | 0.0000 | 0.0000 | 0.0033 | 0.0000 | 0.0000 | 0.0000 | 0.0045 | 0.0089 | 0.0033 | 0.0067 | 0.0022 | 0.0000 | 0.0011 | 0.0000 | 0.0022 | 0.0011 | 0.0011 | 0.0000 | 0.0000  |\n",
      "|   R    | 0.0011 | 0.0056 | 0.0000 | 0.0022 | 0.0067 | 0.0000 | 0.0000 | 0.0022 | 0.0089 | 0.0056 | 0.0000 | 0.0022 | 0.0000 | 0.0011 | 0.0056 | 0.0033 | 0.0033 | 0.0033 | 0.0000 | 0.0022  |\n",
      "|   S    | 0.0033 | 0.0033 | 0.0000 | 0.0011 | 0.0033 | 0.0033 | 0.0000 | 0.0033 | 0.0011 | 0.0022 | 0.0000 | 0.0033 | 0.0022 | 0.0011 | 0.0056 | 0.0011 | 0.0011 | 0.0000 | 0.0011 | 0.0045  |\n",
      "|   T    | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0011 | 0.0022 | 0.0000 | 0.0022 | 0.0033 | 0.0022 | 0.0000 | 0.0022 | 0.0011 | 0.0011 | 0.0000 | 0.0022 | 0.0000 | 0.0045 | 0.0000 | 0.0022  |\n",
      "|   V    | 0.0011 | 0.0022 | 0.0011 | 0.0000 | 0.0022 | 0.0011 | 0.0000 | 0.0011 | 0.0011 | 0.0089 | 0.0011 | 0.0022 | 0.0000 | 0.0022 | 0.0022 | 0.0011 | 0.0011 | 0.0033 | 0.0033 | 0.0000  |\n",
      "|   W    | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0033 | 0.0011 | 0.0011 | 0.0011 | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0000 | 0.0000 | 0.0011 | 0.0000 | 0.0000 | 0.0011 | 0.0011 | 0.0011  |\n",
      "|   Y    | 0.0011 | 0.0000 | 0.0011 | 0.0011 | 0.0033 | 0.0033 | 0.0033 | 0.0022 | 0.0056 | 0.0056 | 0.0011 | 0.0067 | 0.0000 | 0.0011 | 0.0022 | 0.0045 | 0.0011 | 0.0011 | 0.0000 | 0.0033  |\n",
      "----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "print_results(\"data\\\\bacterial4.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "66273372",
   "metadata": {},
   "outputs": [],
   "source": [
    "#aggregate same category viruses to one \n",
    "def aggregate_results_by_category(file_names):\n",
    "    total_amino_acid_counts = {aa: 0 for aa in 'ACDEFGHIKLMNPQRSTVWY'}\n",
    "    total_dikodon_counts = {(a, b): 0 for a in 'ACDEFGHIKLMNPQRSTVWY' for b in 'ACDEFGHIKLMNPQRSTVWY'}\n",
    "\n",
    "    protein_sequences = []\n",
    "\n",
    "    for file_name in file_names:\n",
    "        sequence_record = SeqIO.read(file_name, \"fasta\")\n",
    "        sequence = str(sequence_record.seq)\n",
    "\n",
    "        pairs = extract_valid_start_stop_pairs(sequence)\n",
    "        filtered_fragments = filter_long_fragment_pairs(sequence, pairs)\n",
    "\n",
    "        for start, stop in filtered_fragments:\n",
    "            dna_fragment = sequence[start:stop+3]\n",
    "            protein_sequence = convert_to_protein(dna_fragment)\n",
    "            protein_sequences.append(protein_sequence)\n",
    "\n",
    "            dikodons = extract_dikodons(protein_sequence)\n",
    "            for dikodon in dikodons:\n",
    "                if (dikodon[0], dikodon[1]) in total_dikodon_counts:\n",
    "                    total_dikodon_counts[(dikodon[0], dikodon[1])] += 1\n",
    "\n",
    "    amino_acid_freqs = amino_acid_frequencies(protein_sequences)\n",
    "\n",
    "    total_dikodons = sum(total_dikodon_counts.values())\n",
    "    dikodon_frequencies = {dikodon: count/total_dikodons for dikodon, count in total_dikodon_counts.items()}\n",
    "    \n",
    "    return amino_acid_freqs, dikodon_frequencies\n",
    "\n",
    "#top 15\n",
    "def print_categroty_results(amino_acid_frequencies, dikodon_frequencies):\n",
    "    print(\"Amino Acid | Frequency\")\n",
    "    \n",
    "    sorted_amino_acids = sorted(amino_acid_frequencies.items(), key=lambda x: x[1], reverse=True)\n",
    "    for aa, freq in sorted_amino_acids:\n",
    "        print(f\"{aa} \\t {freq:.5f}\")\n",
    "\n",
    "    print(\"\\nDikodon Frequencies:\")\n",
    "    \n",
    "    sorted_dikodons = sorted(dikodon_frequencies.items(), key=lambda x: x[1], reverse=True)[:15]\n",
    "    for (a, b), freq in sorted_dikodons:\n",
    "        print(f\"({a}, {b}) \\t {freq:.5f}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "9b5efed7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino Acid | Frequency\n",
      "L \t 0.09716\n",
      "M \t 0.09534\n",
      "S \t 0.08899\n",
      "R \t 0.06933\n",
      "V \t 0.06148\n",
      "I \t 0.05707\n",
      "T \t 0.05589\n",
      "F \t 0.04955\n",
      "G \t 0.04955\n",
      "A \t 0.04847\n",
      "P \t 0.04697\n",
      "K \t 0.03708\n",
      "C \t 0.03590\n",
      "Y \t 0.03590\n",
      "Q \t 0.03493\n",
      "H \t 0.03386\n",
      "N \t 0.03353\n",
      "E \t 0.02601\n",
      "D \t 0.02332\n",
      "W \t 0.01967\n",
      "\n",
      "Dikodon Frequencies:\n",
      "(L, L) \t 0.01440\n",
      "(S, S) \t 0.01245\n",
      "(M, L) \t 0.01219\n",
      "(M, S) \t 0.01089\n",
      "(R, S) \t 0.00921\n",
      "(R, R) \t 0.00843\n",
      "(L, S) \t 0.00817\n",
      "(S, L) \t 0.00804\n",
      "(V, L) \t 0.00804\n",
      "(I, L) \t 0.00778\n",
      "(S, R) \t 0.00765\n",
      "(M, V) \t 0.00752\n",
      "(V, S) \t 0.00752\n",
      "(L, P) \t 0.00713\n",
      "(L, F) \t 0.00687\n"
     ]
    }
   ],
   "source": [
    "#mamalian frequencies\n",
    "mammalian_filenames = [\"data\\\\mamalian1.fasta\", \"data\\\\mamalian2.fasta\", \"data\\\\mamalian3.fasta\", \n",
    "               \"data\\\\mamalian4.fasta\"]\n",
    "mam_amino_acid_freqs, mam_dikodon_freqs = aggregate_results_by_category(mammalian_filenames)\n",
    "print_categroty_results(mam_amino_acid_freqs, mam_dikodon_freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "901dd436",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino Acid | Frequency\n",
      "M \t 0.13285\n",
      "L \t 0.10845\n",
      "S \t 0.07763\n",
      "R \t 0.06293\n",
      "I \t 0.06150\n",
      "V \t 0.05908\n",
      "K \t 0.05765\n",
      "F \t 0.04709\n",
      "T \t 0.04466\n",
      "G \t 0.04295\n",
      "A \t 0.04067\n",
      "Q \t 0.04067\n",
      "Y \t 0.03810\n",
      "N \t 0.03696\n",
      "P \t 0.03382\n",
      "C \t 0.02711\n",
      "D \t 0.02469\n",
      "E \t 0.02212\n",
      "H \t 0.02197\n",
      "W \t 0.01912\n",
      "\n",
      "Dikodon Frequencies:\n",
      "(M, L) \t 0.01717\n",
      "(L, L) \t 0.01610\n",
      "(I, L) \t 0.01252\n",
      "(M, K) \t 0.01234\n",
      "(L, S) \t 0.01163\n",
      "(M, I) \t 0.01127\n",
      "(M, V) \t 0.01127\n",
      "(M, S) \t 0.00948\n",
      "(S, S) \t 0.00948\n",
      "(M, F) \t 0.00912\n",
      "(M, Q) \t 0.00877\n",
      "(S, L) \t 0.00877\n",
      "(V, L) \t 0.00859\n",
      "(L, V) \t 0.00841\n",
      "(M, T) \t 0.00841\n"
     ]
    }
   ],
   "source": [
    "#bacterial frequencies\n",
    "bacterial_filenames = [\"data\\\\bacterial1.fasta\", \"data\\\\bacterial2.fasta\",\n",
    "               \"data\\\\bacterial3.fasta\", \"data\\\\bacterial4.fasta\"]\n",
    "bac_amino_acid_freqs, bac_dikodon_freqs = aggregate_results_by_category(bacterial_filenames)\n",
    "print_categroty_results(bac_amino_acid_freqs, bac_dikodon_freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a4df2e52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Amino Acid: M - Difference: 0.0375\n",
      "Amino Acid: K - Difference: 0.0206\n",
      "Amino Acid: P - Difference: 0.0132\n",
      "Amino Acid: H - Difference: 0.0119\n",
      "Amino Acid: S - Difference: 0.0114\n",
      "Amino Acid: L - Difference: 0.0113\n",
      "Amino Acid: T - Difference: 0.0112\n",
      "Amino Acid: C - Difference: 0.0088\n",
      "Amino Acid: A - Difference: 0.0078\n",
      "Amino Acid: G - Difference: 0.0066\n",
      "Amino Acid: R - Difference: 0.0064\n",
      "Amino Acid: Q - Difference: 0.0057\n",
      "Amino Acid: I - Difference: 0.0044\n",
      "Amino Acid: E - Difference: 0.0039\n",
      "Amino Acid: N - Difference: 0.0034\n",
      "Amino Acid: F - Difference: 0.0025\n",
      "Amino Acid: V - Difference: 0.0024\n",
      "Amino Acid: Y - Difference: 0.0022\n",
      "Amino Acid: D - Difference: 0.0014\n",
      "Amino Acid: W - Difference: 0.0005\n"
     ]
    }
   ],
   "source": [
    "def compare_amino_acid_frequencies(mam_amino_acid_freqs, bac_amino_acid_freqs):\n",
    "    amino_acid_difference = {aa: abs(mam_amino_acid_freqs[aa] - bac_amino_acid_freqs[aa]) for aa in mam_amino_acid_freqs}\n",
    "    \n",
    "    sorted_aas = sorted(amino_acid_difference.keys(), key=lambda x: abs(amino_acid_difference[x]), reverse=True)\n",
    "    \n",
    "    for aa in sorted_aas:\n",
    "        difference = amino_acid_difference[aa]\n",
    "        print(f\"Amino Acid: {aa} - Difference: {difference:.4f}\")\n",
    "        \n",
    "compare_amino_acid_frequencies(mam_amino_acid_freqs, bac_amino_acid_freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "502ebba5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dikodon: ('M', 'K') - Difference: 0.0078\n",
      "Dikodon: ('M', 'Q') - Difference: 0.0059\n",
      "Dikodon: ('M', 'F') - Difference: 0.0052\n",
      "Dikodon: ('M', 'L') - Difference: 0.0050\n",
      "Dikodon: ('I', 'L') - Difference: 0.0047\n",
      "Dikodon: ('K', 'R') - Difference: 0.0047\n",
      "Dikodon: ('M', 'I') - Difference: 0.0047\n",
      "Dikodon: ('R', 'S') - Difference: 0.0040\n",
      "Dikodon: ('M', 'V') - Difference: 0.0037\n",
      "Dikodon: ('M', 'T') - Difference: 0.0036\n",
      "Dikodon: ('M', 'D') - Difference: 0.0035\n",
      "Dikodon: ('L', 'S') - Difference: 0.0035\n",
      "Dikodon: ('T', 'S') - Difference: 0.0034\n",
      "Dikodon: ('K', 'L') - Difference: 0.0032\n",
      "Dikodon: ('S', 'R') - Difference: 0.0032\n",
      "Dikodon: ('C', 'L') - Difference: 0.0032\n",
      "Dikodon: ('P', 'R') - Difference: 0.0031\n",
      "Dikodon: ('M', 'A') - Difference: 0.0031\n",
      "Dikodon: ('M', 'Y') - Difference: 0.0031\n",
      "Dikodon: ('H', 'L') - Difference: 0.0031\n",
      "Dikodon: ('R', 'G') - Difference: 0.0030\n",
      "Dikodon: ('M', 'E') - Difference: 0.0030\n",
      "Dikodon: ('K', 'K') - Difference: 0.0030\n",
      "Dikodon: ('S', 'S') - Difference: 0.0030\n",
      "Dikodon: ('R', 'R') - Difference: 0.0029\n",
      "Dikodon: ('M', 'N') - Difference: 0.0029\n",
      "Dikodon: ('R', 'A') - Difference: 0.0028\n",
      "Dikodon: ('S', 'A') - Difference: 0.0027\n",
      "Dikodon: ('G', 'I') - Difference: 0.0027\n",
      "Dikodon: ('G', 'S') - Difference: 0.0027\n",
      "Dikodon: ('S', 'T') - Difference: 0.0026\n",
      "Dikodon: ('P', 'V') - Difference: 0.0025\n",
      "Dikodon: ('V', 'S') - Difference: 0.0025\n",
      "Dikodon: ('G', 'G') - Difference: 0.0025\n",
      "Dikodon: ('E', 'E') - Difference: 0.0024\n",
      "Dikodon: ('L', 'I') - Difference: 0.0024\n",
      "Dikodon: ('K', 'N') - Difference: 0.0023\n",
      "Dikodon: ('L', 'P') - Difference: 0.0023\n",
      "Dikodon: ('S', 'I') - Difference: 0.0022\n",
      "Dikodon: ('A', 'A') - Difference: 0.0022\n",
      "Dikodon: ('G', 'P') - Difference: 0.0022\n",
      "Dikodon: ('L', 'V') - Difference: 0.0022\n",
      "Dikodon: ('M', 'C') - Difference: 0.0022\n",
      "Dikodon: ('R', 'T') - Difference: 0.0021\n",
      "Dikodon: ('F', 'K') - Difference: 0.0021\n",
      "Dikodon: ('T', 'L') - Difference: 0.0021\n",
      "Dikodon: ('Q', 'V') - Difference: 0.0021\n",
      "Dikodon: ('A', 'L') - Difference: 0.0019\n",
      "Dikodon: ('Y', 'L') - Difference: 0.0019\n",
      "Dikodon: ('I', 'R') - Difference: 0.0018\n",
      "Dikodon: ('Y', 'A') - Difference: 0.0018\n",
      "Dikodon: ('P', 'T') - Difference: 0.0018\n",
      "Dikodon: ('F', 'P') - Difference: 0.0018\n",
      "Dikodon: ('Y', 'H') - Difference: 0.0018\n",
      "Dikodon: ('P', 'S') - Difference: 0.0018\n",
      "Dikodon: ('A', 'T') - Difference: 0.0017\n",
      "Dikodon: ('V', 'F') - Difference: 0.0017\n",
      "Dikodon: ('L', 'L') - Difference: 0.0017\n",
      "Dikodon: ('K', 'T') - Difference: 0.0017\n",
      "Dikodon: ('H', 'C') - Difference: 0.0017\n",
      "Dikodon: ('N', 'L') - Difference: 0.0017\n",
      "Dikodon: ('R', 'K') - Difference: 0.0017\n",
      "Dikodon: ('K', 'G') - Difference: 0.0017\n",
      "Dikodon: ('L', 'E') - Difference: 0.0016\n",
      "Dikodon: ('L', 'T') - Difference: 0.0016\n",
      "Dikodon: ('C', 'R') - Difference: 0.0016\n",
      "Dikodon: ('G', 'L') - Difference: 0.0015\n",
      "Dikodon: ('V', 'T') - Difference: 0.0015\n",
      "Dikodon: ('I', 'H') - Difference: 0.0015\n",
      "Dikodon: ('R', 'D') - Difference: 0.0015\n",
      "Dikodon: ('E', 'C') - Difference: 0.0015\n",
      "Dikodon: ('T', 'H') - Difference: 0.0015\n",
      "Dikodon: ('R', 'C') - Difference: 0.0015\n",
      "Dikodon: ('E', 'S') - Difference: 0.0015\n",
      "Dikodon: ('F', 'I') - Difference: 0.0015\n",
      "Dikodon: ('N', 'Y') - Difference: 0.0015\n",
      "Dikodon: ('F', 'D') - Difference: 0.0015\n",
      "Dikodon: ('V', 'Q') - Difference: 0.0015\n",
      "Dikodon: ('A', 'G') - Difference: 0.0015\n",
      "Dikodon: ('T', 'T') - Difference: 0.0014\n",
      "Dikodon: ('V', 'G') - Difference: 0.0014\n",
      "Dikodon: ('F', 'H') - Difference: 0.0014\n",
      "Dikodon: ('Y', 'Q') - Difference: 0.0014\n",
      "Dikodon: ('D', 'P') - Difference: 0.0014\n",
      "Dikodon: ('R', 'W') - Difference: 0.0014\n",
      "Dikodon: ('M', 'S') - Difference: 0.0014\n",
      "Dikodon: ('D', 'S') - Difference: 0.0014\n",
      "Dikodon: ('L', 'D') - Difference: 0.0014\n",
      "Dikodon: ('A', 'P') - Difference: 0.0014\n",
      "Dikodon: ('F', 'V') - Difference: 0.0014\n",
      "Dikodon: ('P', 'I') - Difference: 0.0014\n",
      "Dikodon: ('G', 'C') - Difference: 0.0014\n",
      "Dikodon: ('H', 'I') - Difference: 0.0014\n",
      "Dikodon: ('H', 'H') - Difference: 0.0013\n",
      "Dikodon: ('K', 'V') - Difference: 0.0013\n",
      "Dikodon: ('D', 'F') - Difference: 0.0013\n",
      "Dikodon: ('N', 'A') - Difference: 0.0013\n",
      "Dikodon: ('I', 'I') - Difference: 0.0013\n",
      "Dikodon: ('G', 'E') - Difference: 0.0013\n",
      "Dikodon: ('Y', 'T') - Difference: 0.0013\n",
      "Dikodon: ('L', 'W') - Difference: 0.0013\n",
      "Dikodon: ('M', 'H') - Difference: 0.0013\n",
      "Dikodon: ('S', 'Q') - Difference: 0.0013\n",
      "Dikodon: ('Q', 'K') - Difference: 0.0012\n",
      "Dikodon: ('T', 'N') - Difference: 0.0012\n",
      "Dikodon: ('T', 'W') - Difference: 0.0012\n",
      "Dikodon: ('F', 'L') - Difference: 0.0012\n",
      "Dikodon: ('G', 'D') - Difference: 0.0012\n",
      "Dikodon: ('N', 'V') - Difference: 0.0012\n",
      "Dikodon: ('F', 'A') - Difference: 0.0012\n",
      "Dikodon: ('N', 'K') - Difference: 0.0012\n",
      "Dikodon: ('F', 'F') - Difference: 0.0012\n",
      "Dikodon: ('G', 'A') - Difference: 0.0012\n",
      "Dikodon: ('H', 'R') - Difference: 0.0012\n",
      "Dikodon: ('P', 'F') - Difference: 0.0012\n",
      "Dikodon: ('M', 'W') - Difference: 0.0012\n",
      "Dikodon: ('P', 'Y') - Difference: 0.0012\n",
      "Dikodon: ('E', 'L') - Difference: 0.0012\n",
      "Dikodon: ('W', 'S') - Difference: 0.0012\n",
      "Dikodon: ('Q', 'F') - Difference: 0.0011\n",
      "Dikodon: ('F', 'Q') - Difference: 0.0011\n",
      "Dikodon: ('F', 'T') - Difference: 0.0011\n",
      "Dikodon: ('I', 'C') - Difference: 0.0011\n",
      "Dikodon: ('P', 'P') - Difference: 0.0011\n",
      "Dikodon: ('R', 'P') - Difference: 0.0011\n",
      "Dikodon: ('V', 'P') - Difference: 0.0011\n",
      "Dikodon: ('C', 'K') - Difference: 0.0011\n",
      "Dikodon: ('G', 'K') - Difference: 0.0011\n",
      "Dikodon: ('A', 'V') - Difference: 0.0011\n",
      "Dikodon: ('C', 'T') - Difference: 0.0011\n",
      "Dikodon: ('G', 'H') - Difference: 0.0011\n",
      "Dikodon: ('V', 'E') - Difference: 0.0011\n",
      "Dikodon: ('V', 'R') - Difference: 0.0011\n",
      "Dikodon: ('C', 'N') - Difference: 0.0011\n",
      "Dikodon: ('Q', 'H') - Difference: 0.0011\n",
      "Dikodon: ('A', 'W') - Difference: 0.0011\n",
      "Dikodon: ('C', 'E') - Difference: 0.0011\n",
      "Dikodon: ('W', 'C') - Difference: 0.0011\n",
      "Dikodon: ('L', 'M') - Difference: 0.0011\n",
      "Dikodon: ('V', 'W') - Difference: 0.0011\n",
      "Dikodon: ('L', 'G') - Difference: 0.0011\n",
      "Dikodon: ('N', 'P') - Difference: 0.0011\n",
      "Dikodon: ('I', 'F') - Difference: 0.0010\n",
      "Dikodon: ('M', 'G') - Difference: 0.0010\n",
      "Dikodon: ('T', 'K') - Difference: 0.0010\n",
      "Dikodon: ('L', 'C') - Difference: 0.0010\n",
      "Dikodon: ('Q', 'I') - Difference: 0.0010\n",
      "Dikodon: ('D', 'A') - Difference: 0.0010\n",
      "Dikodon: ('P', 'G') - Difference: 0.0010\n",
      "Dikodon: ('T', 'R') - Difference: 0.0010\n",
      "Dikodon: ('Q', 'M') - Difference: 0.0010\n",
      "Dikodon: ('I', 'Y') - Difference: 0.0010\n",
      "Dikodon: ('E', 'P') - Difference: 0.0010\n",
      "Dikodon: ('Q', 'D') - Difference: 0.0009\n",
      "Dikodon: ('I', 'N') - Difference: 0.0009\n",
      "Dikodon: ('V', 'N') - Difference: 0.0009\n",
      "Dikodon: ('H', 'Q') - Difference: 0.0009\n",
      "Dikodon: ('M', 'M') - Difference: 0.0009\n",
      "Dikodon: ('H', 'G') - Difference: 0.0009\n",
      "Dikodon: ('Y', 'G') - Difference: 0.0009\n",
      "Dikodon: ('C', 'A') - Difference: 0.0009\n",
      "Dikodon: ('R', 'Q') - Difference: 0.0009\n",
      "Dikodon: ('G', 'Q') - Difference: 0.0009\n",
      "Dikodon: ('K', 'M') - Difference: 0.0009\n",
      "Dikodon: ('R', 'E') - Difference: 0.0009\n",
      "Dikodon: ('A', 'K') - Difference: 0.0009\n",
      "Dikodon: ('Q', 'Y') - Difference: 0.0008\n",
      "Dikodon: ('S', 'P') - Difference: 0.0008\n",
      "Dikodon: ('F', 'C') - Difference: 0.0008\n",
      "Dikodon: ('G', 'Y') - Difference: 0.0008\n",
      "Dikodon: ('Q', 'T') - Difference: 0.0008\n",
      "Dikodon: ('T', 'A') - Difference: 0.0008\n",
      "Dikodon: ('H', 'N') - Difference: 0.0008\n",
      "Dikodon: ('S', 'V') - Difference: 0.0008\n",
      "Dikodon: ('C', 'P') - Difference: 0.0008\n",
      "Dikodon: ('H', 'Y') - Difference: 0.0008\n",
      "Dikodon: ('D', 'I') - Difference: 0.0008\n",
      "Dikodon: ('F', 'G') - Difference: 0.0008\n",
      "Dikodon: ('E', 'K') - Difference: 0.0008\n",
      "Dikodon: ('V', 'M') - Difference: 0.0008\n",
      "Dikodon: ('G', 'F') - Difference: 0.0008\n",
      "Dikodon: ('N', 'T') - Difference: 0.0008\n",
      "Dikodon: ('N', 'F') - Difference: 0.0008\n",
      "Dikodon: ('C', 'H') - Difference: 0.0008\n",
      "Dikodon: ('H', 'T') - Difference: 0.0008\n",
      "Dikodon: ('E', 'Q') - Difference: 0.0008\n",
      "Dikodon: ('K', 'H') - Difference: 0.0007\n",
      "Dikodon: ('Q', 'G') - Difference: 0.0007\n",
      "Dikodon: ('V', 'C') - Difference: 0.0007\n",
      "Dikodon: ('Y', 'R') - Difference: 0.0007\n",
      "Dikodon: ('T', 'G') - Difference: 0.0007\n",
      "Dikodon: ('V', 'I') - Difference: 0.0007\n",
      "Dikodon: ('S', 'L') - Difference: 0.0007\n",
      "Dikodon: ('Y', 'K') - Difference: 0.0007\n",
      "Dikodon: ('K', 'A') - Difference: 0.0007\n",
      "Dikodon: ('P', 'H') - Difference: 0.0007\n",
      "Dikodon: ('T', 'E') - Difference: 0.0007\n",
      "Dikodon: ('G', 'V') - Difference: 0.0007\n",
      "Dikodon: ('R', 'I') - Difference: 0.0007\n",
      "Dikodon: ('D', 'D') - Difference: 0.0007\n",
      "Dikodon: ('Q', 'N') - Difference: 0.0007\n",
      "Dikodon: ('F', 'R') - Difference: 0.0007\n",
      "Dikodon: ('T', 'M') - Difference: 0.0007\n",
      "Dikodon: ('Y', 'V') - Difference: 0.0007\n",
      "Dikodon: ('Y', 'I') - Difference: 0.0007\n",
      "Dikodon: ('L', 'Y') - Difference: 0.0007\n",
      "Dikodon: ('I', 'S') - Difference: 0.0006\n",
      "Dikodon: ('I', 'A') - Difference: 0.0006\n",
      "Dikodon: ('V', 'D') - Difference: 0.0006\n",
      "Dikodon: ('L', 'R') - Difference: 0.0006\n",
      "Dikodon: ('I', 'Q') - Difference: 0.0006\n",
      "Dikodon: ('A', 'I') - Difference: 0.0006\n",
      "Dikodon: ('G', 'T') - Difference: 0.0006\n",
      "Dikodon: ('T', 'V') - Difference: 0.0006\n",
      "Dikodon: ('C', 'G') - Difference: 0.0006\n",
      "Dikodon: ('K', 'E') - Difference: 0.0006\n",
      "Dikodon: ('P', 'D') - Difference: 0.0006\n",
      "Dikodon: ('L', 'K') - Difference: 0.0006\n",
      "Dikodon: ('N', 'I') - Difference: 0.0006\n",
      "Dikodon: ('I', 'V') - Difference: 0.0006\n",
      "Dikodon: ('C', 'M') - Difference: 0.0006\n",
      "Dikodon: ('H', 'D') - Difference: 0.0006\n",
      "Dikodon: ('K', 'I') - Difference: 0.0006\n",
      "Dikodon: ('I', 'P') - Difference: 0.0006\n",
      "Dikodon: ('W', 'I') - Difference: 0.0006\n",
      "Dikodon: ('M', 'R') - Difference: 0.0006\n",
      "Dikodon: ('C', 'V') - Difference: 0.0006\n",
      "Dikodon: ('T', 'P') - Difference: 0.0006\n",
      "Dikodon: ('E', 'N') - Difference: 0.0006\n",
      "Dikodon: ('S', 'C') - Difference: 0.0006\n",
      "Dikodon: ('N', 'E') - Difference: 0.0006\n",
      "Dikodon: ('W', 'A') - Difference: 0.0006\n",
      "Dikodon: ('D', 'H') - Difference: 0.0006\n",
      "Dikodon: ('E', 'D') - Difference: 0.0006\n",
      "Dikodon: ('I', 'D') - Difference: 0.0006\n",
      "Dikodon: ('V', 'L') - Difference: 0.0005\n",
      "Dikodon: ('K', 'S') - Difference: 0.0005\n",
      "Dikodon: ('K', 'P') - Difference: 0.0005\n",
      "Dikodon: ('T', 'Y') - Difference: 0.0005\n",
      "Dikodon: ('G', 'R') - Difference: 0.0005\n",
      "Dikodon: ('L', 'A') - Difference: 0.0005\n",
      "Dikodon: ('P', 'Q') - Difference: 0.0005\n",
      "Dikodon: ('V', 'H') - Difference: 0.0005\n",
      "Dikodon: ('R', 'L') - Difference: 0.0005\n",
      "Dikodon: ('K', 'C') - Difference: 0.0005\n",
      "Dikodon: ('A', 'C') - Difference: 0.0005\n",
      "Dikodon: ('D', 'V') - Difference: 0.0005\n",
      "Dikodon: ('N', 'G') - Difference: 0.0005\n",
      "Dikodon: ('W', 'H') - Difference: 0.0005\n",
      "Dikodon: ('A', 'H') - Difference: 0.0005\n",
      "Dikodon: ('E', 'A') - Difference: 0.0005\n",
      "Dikodon: ('I', 'E') - Difference: 0.0005\n",
      "Dikodon: ('Y', 'F') - Difference: 0.0005\n",
      "Dikodon: ('P', 'C') - Difference: 0.0005\n",
      "Dikodon: ('P', 'N') - Difference: 0.0005\n",
      "Dikodon: ('N', 'R') - Difference: 0.0005\n",
      "Dikodon: ('H', 'K') - Difference: 0.0005\n",
      "Dikodon: ('P', 'W') - Difference: 0.0005\n",
      "Dikodon: ('I', 'G') - Difference: 0.0005\n",
      "Dikodon: ('D', 'N') - Difference: 0.0005\n",
      "Dikodon: ('W', 'N') - Difference: 0.0005\n",
      "Dikodon: ('V', 'Y') - Difference: 0.0004\n",
      "Dikodon: ('N', 'D') - Difference: 0.0004\n",
      "Dikodon: ('F', 'E') - Difference: 0.0004\n",
      "Dikodon: ('Y', 'W') - Difference: 0.0004\n",
      "Dikodon: ('D', 'C') - Difference: 0.0004\n",
      "Dikodon: ('W', 'E') - Difference: 0.0004\n",
      "Dikodon: ('W', 'M') - Difference: 0.0004\n",
      "Dikodon: ('Y', 'S') - Difference: 0.0004\n",
      "Dikodon: ('T', 'Q') - Difference: 0.0004\n",
      "Dikodon: ('E', 'T') - Difference: 0.0004\n",
      "Dikodon: ('A', 'Y') - Difference: 0.0004\n",
      "Dikodon: ('F', 'M') - Difference: 0.0004\n",
      "Dikodon: ('Y', 'P') - Difference: 0.0004\n",
      "Dikodon: ('G', 'M') - Difference: 0.0004\n",
      "Dikodon: ('H', 'M') - Difference: 0.0004\n",
      "Dikodon: ('I', 'M') - Difference: 0.0004\n",
      "Dikodon: ('K', 'W') - Difference: 0.0004\n",
      "Dikodon: ('R', 'V') - Difference: 0.0004\n",
      "Dikodon: ('T', 'D') - Difference: 0.0004\n",
      "Dikodon: ('C', 'S') - Difference: 0.0004\n",
      "Dikodon: ('R', 'Y') - Difference: 0.0004\n",
      "Dikodon: ('G', 'N') - Difference: 0.0004\n",
      "Dikodon: ('C', 'C') - Difference: 0.0004\n",
      "Dikodon: ('I', 'W') - Difference: 0.0004\n",
      "Dikodon: ('N', 'H') - Difference: 0.0004\n",
      "Dikodon: ('Q', 'E') - Difference: 0.0004\n",
      "Dikodon: ('W', 'T') - Difference: 0.0004\n",
      "Dikodon: ('Y', 'C') - Difference: 0.0004\n",
      "Dikodon: ('A', 'R') - Difference: 0.0003\n",
      "Dikodon: ('C', 'I') - Difference: 0.0003\n",
      "Dikodon: ('D', 'R') - Difference: 0.0003\n",
      "Dikodon: ('F', 'Y') - Difference: 0.0003\n",
      "Dikodon: ('V', 'V') - Difference: 0.0003\n",
      "Dikodon: ('E', 'H') - Difference: 0.0003\n",
      "Dikodon: ('H', 'W') - Difference: 0.0003\n",
      "Dikodon: ('H', 'A') - Difference: 0.0003\n",
      "Dikodon: ('S', 'H') - Difference: 0.0003\n",
      "Dikodon: ('D', 'E') - Difference: 0.0003\n",
      "Dikodon: ('Q', 'W') - Difference: 0.0003\n",
      "Dikodon: ('W', 'F') - Difference: 0.0003\n",
      "Dikodon: ('R', 'N') - Difference: 0.0003\n",
      "Dikodon: ('V', 'A') - Difference: 0.0003\n",
      "Dikodon: ('Q', 'S') - Difference: 0.0003\n",
      "Dikodon: ('N', 'N') - Difference: 0.0003\n",
      "Dikodon: ('P', 'E') - Difference: 0.0003\n",
      "Dikodon: ('K', 'F') - Difference: 0.0003\n",
      "Dikodon: ('K', 'D') - Difference: 0.0003\n",
      "Dikodon: ('G', 'W') - Difference: 0.0003\n",
      "Dikodon: ('A', 'E') - Difference: 0.0003\n",
      "Dikodon: ('Y', 'Y') - Difference: 0.0003\n",
      "Dikodon: ('L', 'F') - Difference: 0.0003\n",
      "Dikodon: ('D', 'Q') - Difference: 0.0003\n",
      "Dikodon: ('H', 'E') - Difference: 0.0003\n",
      "Dikodon: ('A', 'D') - Difference: 0.0003\n",
      "Dikodon: ('F', 'W') - Difference: 0.0003\n",
      "Dikodon: ('K', 'Q') - Difference: 0.0003\n",
      "Dikodon: ('W', 'Y') - Difference: 0.0003\n",
      "Dikodon: ('N', 'S') - Difference: 0.0003\n",
      "Dikodon: ('R', 'H') - Difference: 0.0003\n",
      "Dikodon: ('S', 'N') - Difference: 0.0003\n",
      "Dikodon: ('R', 'F') - Difference: 0.0003\n",
      "Dikodon: ('S', 'Y') - Difference: 0.0003\n",
      "Dikodon: ('D', 'M') - Difference: 0.0002\n",
      "Dikodon: ('P', 'M') - Difference: 0.0002\n",
      "Dikodon: ('Y', 'E') - Difference: 0.0002\n",
      "Dikodon: ('S', 'F') - Difference: 0.0002\n",
      "Dikodon: ('W', 'R') - Difference: 0.0002\n",
      "Dikodon: ('C', 'Y') - Difference: 0.0002\n",
      "Dikodon: ('W', 'V') - Difference: 0.0002\n",
      "Dikodon: ('S', 'K') - Difference: 0.0002\n",
      "Dikodon: ('C', 'Q') - Difference: 0.0002\n",
      "Dikodon: ('N', 'C') - Difference: 0.0002\n",
      "Dikodon: ('S', 'G') - Difference: 0.0002\n",
      "Dikodon: ('H', 'F') - Difference: 0.0002\n",
      "Dikodon: ('I', 'K') - Difference: 0.0002\n",
      "Dikodon: ('T', 'C') - Difference: 0.0002\n",
      "Dikodon: ('W', 'D') - Difference: 0.0002\n",
      "Dikodon: ('W', 'G') - Difference: 0.0002\n",
      "Dikodon: ('H', 'V') - Difference: 0.0002\n",
      "Dikodon: ('E', 'R') - Difference: 0.0002\n",
      "Dikodon: ('Y', 'M') - Difference: 0.0002\n",
      "Dikodon: ('Q', 'R') - Difference: 0.0002\n",
      "Dikodon: ('N', 'Q') - Difference: 0.0001\n",
      "Dikodon: ('N', 'W') - Difference: 0.0001\n",
      "Dikodon: ('W', 'P') - Difference: 0.0001\n",
      "Dikodon: ('D', 'K') - Difference: 0.0001\n",
      "Dikodon: ('Q', 'C') - Difference: 0.0001\n",
      "Dikodon: ('H', 'S') - Difference: 0.0001\n",
      "Dikodon: ('A', 'F') - Difference: 0.0001\n",
      "Dikodon: ('E', 'I') - Difference: 0.0001\n",
      "Dikodon: ('K', 'Y') - Difference: 0.0001\n",
      "Dikodon: ('Q', 'P') - Difference: 0.0001\n",
      "Dikodon: ('W', 'K') - Difference: 0.0001\n",
      "Dikodon: ('S', 'D') - Difference: 0.0001\n",
      "Dikodon: ('D', 'L') - Difference: 0.0001\n",
      "Dikodon: ('C', 'W') - Difference: 0.0001\n",
      "Dikodon: ('F', 'S') - Difference: 0.0001\n",
      "Dikodon: ('M', 'P') - Difference: 0.0001\n",
      "Dikodon: ('R', 'M') - Difference: 0.0001\n",
      "Dikodon: ('S', 'M') - Difference: 0.0001\n",
      "Dikodon: ('T', 'F') - Difference: 0.0001\n",
      "Dikodon: ('A', 'Q') - Difference: 0.0001\n",
      "Dikodon: ('E', 'V') - Difference: 0.0001\n",
      "Dikodon: ('S', 'E') - Difference: 0.0001\n",
      "Dikodon: ('V', 'K') - Difference: 0.0001\n",
      "Dikodon: ('W', 'L') - Difference: 0.0001\n",
      "Dikodon: ('Y', 'N') - Difference: 0.0001\n",
      "Dikodon: ('Q', 'L') - Difference: 0.0001\n",
      "Dikodon: ('A', 'S') - Difference: 0.0001\n",
      "Dikodon: ('P', 'K') - Difference: 0.0001\n",
      "Dikodon: ('S', 'W') - Difference: 0.0001\n",
      "Dikodon: ('E', 'Y') - Difference: 0.0001\n",
      "Dikodon: ('C', 'F') - Difference: 0.0001\n",
      "Dikodon: ('P', 'A') - Difference: 0.0001\n",
      "Dikodon: ('T', 'I') - Difference: 0.0001\n",
      "Dikodon: ('D', 'W') - Difference: 0.0001\n",
      "Dikodon: ('E', 'F') - Difference: 0.0001\n",
      "Dikodon: ('E', 'G') - Difference: 0.0001\n",
      "Dikodon: ('F', 'N') - Difference: 0.0001\n",
      "Dikodon: ('H', 'P') - Difference: 0.0001\n",
      "Dikodon: ('L', 'N') - Difference: 0.0001\n",
      "Dikodon: ('E', 'W') - Difference: 0.0000\n",
      "Dikodon: ('I', 'T') - Difference: 0.0000\n",
      "Dikodon: ('D', 'G') - Difference: 0.0000\n",
      "Dikodon: ('A', 'M') - Difference: 0.0000\n",
      "Dikodon: ('D', 'T') - Difference: 0.0000\n",
      "Dikodon: ('D', 'Y') - Difference: 0.0000\n",
      "Dikodon: ('E', 'M') - Difference: 0.0000\n",
      "Dikodon: ('P', 'L') - Difference: 0.0000\n",
      "Dikodon: ('L', 'Q') - Difference: 0.0000\n",
      "Dikodon: ('A', 'N') - Difference: 0.0000\n",
      "Dikodon: ('Q', 'A') - Difference: 0.0000\n",
      "Dikodon: ('Q', 'Q') - Difference: 0.0000\n",
      "Dikodon: ('C', 'D') - Difference: 0.0000\n",
      "Dikodon: ('N', 'M') - Difference: 0.0000\n",
      "Dikodon: ('W', 'Q') - Difference: 0.0000\n",
      "Dikodon: ('W', 'W') - Difference: 0.0000\n",
      "Dikodon: ('Y', 'D') - Difference: 0.0000\n",
      "Dikodon: ('L', 'H') - Difference: 0.0000\n"
     ]
    }
   ],
   "source": [
    "def compare_dikodon_frequencies(mam_dikodon_freqs, bac_dikodon_freqs):\n",
    "    dikodon_difference = {dk: abs(mam_dikodon_freqs[dk] - bac_dikodon_freqs[dk]) for dk in mam_dikodon_freqs}\n",
    "    \n",
    "    sorted_dks = sorted(dikodon_difference.keys(), key=lambda x: abs(dikodon_difference[x]), reverse=True)\n",
    "    \n",
    "    for dk in sorted_dks:\n",
    "        difference = dikodon_difference[dk]\n",
    "        print(f\"Dikodon: {dk} - Difference: {difference:.4f}\")\n",
    "\n",
    "compare_dikodon_frequencies(mam_dikodon_freqs, bac_dikodon_freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b636a130",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
